
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>autoprot.analysis &#8212; autoprot  documentation</title>
  <script>
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=9b1a4fa89bdd0e95b23b" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=9b1a4fa89bdd0e95b23b" rel="stylesheet">

  
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=9b1a4fa89bdd0e95b23b">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/autoprot/analysis';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fas fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../../index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">autoprot  documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                <li class="nav-item">
                    <a class="nav-link" href="../../installation.html">
                        Installation
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../preprocessing.html">
                        Data Preprocessing
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../analysis.html">
                        Analysis
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../clustering.html">
                        Clustering
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../visualization.html">
                        Data Visualization
                    </a>
                </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                <li class="nav-item">
                    <a class="nav-link" href="../../workflows.html">
                        Workflows
                    </a>
                </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      <div class="navbar-end-item navbar-end__search-button-container">
        
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search">
  <i class="fas fa-search"></i>
</button>
      </div>
      
      <div class="navbar-end-item">
        <span class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle">
    <a class="theme-switch" data-mode="light"><i class="fas fa-sun"></i></a>
    <a class="theme-switch" data-mode="dark"><i class="far fa-moon"></i></a>
    <a class="theme-switch" data-mode="auto"><i class="fas fa-adjust"></i></a>
</span>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  <div class="search-button-container--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search">
  <i class="fas fa-search"></i>
</button>
  </div>

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fas fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                <li class="nav-item">
                    <a class="nav-link" href="../../installation.html">
                        Installation
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../preprocessing.html">
                        Data Preprocessing
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../analysis.html">
                        Analysis
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../clustering.html">
                        Clustering
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="../../visualization.html">
                        Data Visualization
                    </a>
                </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                <li class="nav-item">
                    <a class="nav-link" href="../../workflows.html">
                        Workflows
                    </a>
                </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <span class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle">
    <a class="theme-switch" data-mode="light"><i class="fas fa-sun"></i></a>
    <a class="theme-switch" data-mode="dark"><i class="far fa-moon"></i></a>
    <a class="theme-switch" data-mode="auto"><i class="fas fa-adjust"></i></a>
</span>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

      </div>
      <main class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <h1>Source code for autoprot.analysis</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Autoprot Analysis Functions.</span>

<span class="sd">@author: Wignand</span>

<span class="sd">@documentation: Julian</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">subprocess</span> <span class="kn">import</span> <span class="n">run</span><span class="p">,</span> <span class="n">PIPE</span><span class="p">,</span> <span class="n">STDOUT</span>
<span class="kn">from</span> <span class="nn">importlib</span> <span class="kn">import</span> <span class="n">resources</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Any</span>

<span class="kn">from</span> <span class="nn">statsmodels.stats</span> <span class="kn">import</span> <span class="n">multitest</span> <span class="k">as</span> <span class="n">mt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ttest_1samp</span><span class="p">,</span> <span class="n">ttest_ind</span><span class="p">,</span> <span class="n">wilcoxon</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">import</span> <span class="nn">matplotlib.pylab</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pylab</span> <span class="k">as</span> <span class="nn">pl</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">from</span> <span class="nn">operator</span> <span class="kn">import</span> <span class="n">itemgetter</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="c1"># noinspection PyUnresolvedReferences</span>
<span class="kn">from</span> <span class="nn">autoprot</span> <span class="kn">import</span> <span class="n">visualization</span> <span class="k">as</span> <span class="n">vis</span>
<span class="c1"># noinspection PyUnresolvedReferences</span>
<span class="kn">from</span> <span class="nn">autoprot</span> <span class="kn">import</span> <span class="n">r_helper</span>
<span class="c1"># noinspection PyUnresolvedReferences</span>
<span class="kn">from</span> <span class="nn">autoprot</span> <span class="kn">import</span> <span class="n">preprocessing</span> <span class="k">as</span> <span class="n">pp</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">missingno</span> <span class="k">as</span> <span class="nn">msn</span>
<span class="kn">from</span> <span class="nn">gprofiler</span> <span class="kn">import</span> <span class="n">GProfiler</span>

<span class="n">gp</span> <span class="o">=</span> <span class="n">GProfiler</span><span class="p">(</span>
    <span class="n">user_agent</span><span class="o">=</span><span class="s2">&quot;autoprot&quot;</span><span class="p">,</span>
    <span class="n">return_dataframe</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">RFUNCTIONS</span><span class="p">,</span> <span class="n">R</span> <span class="o">=</span> <span class="n">r_helper</span><span class="o">.</span><span class="n">return_r_path</span><span class="p">()</span>

<span class="c1"># check where this is actually used and make it local</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">diverging_palette</span><span class="p">(</span><span class="mi">150</span><span class="p">,</span> <span class="mi">275</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">l</span><span class="o">=</span><span class="mi">55</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>


<div class="viewcode-block" id="ttest"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.ttest">[docs]</a><span class="k">def</span> <span class="nf">ttest</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">reps</span><span class="p">,</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">return_fc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">adjust_p_vals</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;two-sided&#39;</span><span class="p">,</span> <span class="n">logged</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform one or two sample ttest.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    reps : list of str or list of lists of str</span>
<span class="sd">        The replicates to be included in the statistical test.</span>
<span class="sd">        Either a list of the replicates or a list containing two list with the</span>
<span class="sd">        respective replicates.</span>
<span class="sd">    cond : str, optional</span>
<span class="sd">        The name of the condition.</span>
<span class="sd">        This is used for naming the returned results.</span>
<span class="sd">        The default is &quot;&quot;.</span>
<span class="sd">    return_fc : bool, optional</span>
<span class="sd">        Whether to calculate the fold-change of the provided data.</span>
<span class="sd">        The processing of the fold-change can be controlled by the logged</span>
<span class="sd">        switch.</span>
<span class="sd">        The default is True.</span>
<span class="sd">    adjust_p_vals : bool, optional</span>
<span class="sd">        Whether to adjust P-values. The default is True.</span>
<span class="sd">    alternative : {&#39;two-sided&#39;, &#39;less&#39;, &#39;greater&#39;}, optional</span>
<span class="sd">        Defines the alternative hypothesis.</span>
<span class="sd">        The following options are available (default is &#39;two-sided&#39;):</span>

<span class="sd">        * &#39;two-sided&#39;: the mean of the underlying distribution of the sample</span>
<span class="sd">          is different from the given population mean (`popmean`)</span>
<span class="sd">        * &#39;less&#39;: the mean of the underlying distribution of the sample is</span>
<span class="sd">          less than the given population mean (`popmean`)</span>
<span class="sd">        * &#39;greater&#39;: the mean of the underlying distribution of the sample is</span>
<span class="sd">          greater than the given population mean (`popmean`)</span>
<span class="sd">    logged: bool, optional</span>
<span class="sd">        Set to True if input values are log-transformed (or VSN normalised).</span>
<span class="sd">        This returns the difference between values as log_fc, otherwise</span>
<span class="sd">        values are log2 transformed to gain log_fc.</span>
<span class="sd">        Default is true.</span>


<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe with additional cols.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; twitchVsmild = [&#39;log2_Ratio H/M normalized BC18_1&#39;,&#39;log2_Ratio M/L normalized BC18_2&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio H/M normalized BC18_3&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio H/L normalized BC36_1&#39;,&#39;log2_Ratio H/M normalized BC36_2&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio M/L normalized BC36_2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; protRatio = prot.filter(regex=&quot;Ratio .\/. normalized&quot;)</span>
<span class="sd">    &gt;&gt;&gt; protLog = autoprot.preprocessing.log(prot, protRatio, base=2)</span>
<span class="sd">    &gt;&gt;&gt; prot_tt = autoprot.analysis.ttest(df=protLog, reps=twitchVsmild, cond=&quot;_TvM&quot;, return_fc=True, adjust_p_vals=True)</span>
<span class="sd">    &gt;&gt;&gt; prot_tt[&quot;pValue_TvM&quot;].hist(bins=50)</span>
<span class="sd">    &gt;&gt;&gt; plt.show()</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import autoprot.preprocessing as pp</span>
<span class="sd">        import pandas as pd</span>
<span class="sd">        twitchVsmild = [&#39;log2_Ratio H/M normalized BC18_1&#39;,&#39;log2_Ratio M/L normalized BC18_2&#39;,</span>
<span class="sd">                        &#39;log2_Ratio H/M normalized BC18_3&#39;,</span>
<span class="sd">                        &#39;log2_Ratio H/L normalized BC36_1&#39;,&#39;log2_Ratio H/M normalized BC36_2&#39;,</span>
<span class="sd">                        &#39;log2_Ratio M/L normalized BC36_2&#39;]</span>
<span class="sd">        prot = pd.read_csv(&quot;_static/testdata/proteinGroups.zip&quot;, sep=&#39;\\t&#39;, low_memory=False)</span>
<span class="sd">        protRatio = prot.filter(regex=&quot;Ratio .\/. normalized&quot;)</span>
<span class="sd">        protLog = pp.log(prot, protRatio, base=2)</span>
<span class="sd">        prot_tt = ana.ttest(df=protLog, reps=twitchVsmild, cond=&quot;_TvM&quot;, return_fc=True, adjust_p_vals=True)</span>
<span class="sd">        prot_tt[&quot;pValue_TvM&quot;].hist(bins=50)</span>
<span class="sd">        plt.show()</span>

<span class="sd">    &gt;&gt;&gt; dataframe = pd.DataFrame({&quot;a1&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                           &quot;a2&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                           &quot;a3&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                           &quot;b1&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">    ...                           &quot;b2&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">    ...                           &quot;b3&quot;:np.random.normal(loc=0.5, size=4000),})</span>
<span class="sd">    &gt;&gt;&gt; autoprot.analysis.ttest(df=dataframe, reps=[[&quot;a1&quot;,&quot;a2&quot;, &quot;a3&quot;],[&quot;b1&quot;,&quot;b2&quot;, &quot;b3&quot;]])[&quot;pValue&quot;].hist(bins=50)</span>
<span class="sd">    &gt;&gt;&gt; plt.show()</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import pandas as pd</span>
<span class="sd">        df = pd.DataFrame({&quot;a1&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                  &quot;a2&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                  &quot;a3&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                  &quot;b1&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">                  &quot;b2&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">                  &quot;b3&quot;:np.random.normal(loc=0.5, size=4000),})</span>
<span class="sd">        ana.ttest(df=df, reps=[[&quot;a1&quot;,&quot;a2&quot;, &quot;a3&quot;],[&quot;b1&quot;,&quot;b2&quot;, &quot;b3&quot;]])[&quot;pValue&quot;].hist(bins=50)</span>
<span class="sd">        plt.show()</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">one_samp_ttest</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">filled</span><span class="p">(</span><span class="n">ttest_1samp</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">nan_policy</span><span class="o">=</span><span class="s2">&quot;omit&quot;</span><span class="p">,</span> <span class="n">alternative</span><span class="o">=</span><span class="n">alternative</span><span class="p">,</span> <span class="n">popmean</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">two_samp_ttest</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">filled</span><span class="p">(</span>
            <span class="n">ttest_ind</span><span class="p">(</span><span class="n">x</span><span class="p">[:</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">])],</span> <span class="n">x</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]):],</span> <span class="n">alternative</span><span class="o">=</span><span class="n">alternative</span><span class="p">,</span> <span class="n">nan_policy</span><span class="o">=</span><span class="s2">&quot;omit&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Performing two-sample t-Test&quot;</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pValue</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">two_samp_ttest</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;score</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pValue</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">return_fc</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">logged</span><span class="p">:</span>
                <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;logFC</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;logFC</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">values</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Performing one-sample t-Test&quot;</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pValue</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">one_samp_ttest</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

        <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;score</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;pValue</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">return_fc</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;logFC</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">logged</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">adjust_p_vals</span><span class="p">:</span>
        <span class="n">adjust_p</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pValue</span><span class="si">{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="adjust_p"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.adjust_p">[docs]</a><span class="k">def</span> <span class="nf">adjust_p</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">p_col</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;fdr_bh&quot;</span><span class="p">):</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Use statsmodels.multitest on dataframes.</span>

<span class="sd">    Note: when nan in p-value this function will return only nan.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    p_col : str</span>
<span class="sd">        column containing p-values for correction.</span>
<span class="sd">    method : str, optional</span>
<span class="sd">        &#39;b&#39;: &#39;Bonferroni&#39;,</span>
<span class="sd">        &#39;s&#39;: &#39;Sidak&#39;,</span>
<span class="sd">        &#39;h&#39;: &#39;Holm&#39;,</span>
<span class="sd">        &#39;hs&#39;: &#39;Holm-Sidak&#39;,</span>
<span class="sd">        &#39;sh&#39;: &#39;Simes-Hochberg&#39;,</span>
<span class="sd">        &#39;ho&#39;: &#39;Hommel&#39;,</span>
<span class="sd">        &#39;fdr_bh&#39;: &#39;FDR Benjamini-Hochberg&#39;,</span>
<span class="sd">        &#39;fdr_by&#39;: &#39;FDR Benjamini-Yekutieli&#39;,</span>
<span class="sd">        &#39;fdr_tsbh&#39;: &#39;FDR 2-stage Benjamini-Hochberg&#39;,</span>
<span class="sd">        &#39;fdr_tsbky&#39;: &#39;FDR 2-stage Benjamini-Krieger-Yekutieli&#39;,</span>
<span class="sd">        &#39;fdr_gbs&#39;: &#39;FDR adaptive Gavrilov-Benjamini-Sarkar&#39;</span>
<span class="sd">        The default is &quot;fdr_bh&quot;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        The input dataframe with adjusted p-values.</span>
<span class="sd">        The dataframe will have a column named &quot;adj.{pCol}&quot;</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; twitchVsmild = [&#39;log2_Ratio H/M normalized BC18_1&#39;,&#39;log2_Ratio M/L normalized BC18_2&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio H/M normalized BC18_3&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio H/L normalized BC36_1&#39;,&#39;log2_Ratio H/M normalized BC36_2&#39;,</span>
<span class="sd">    ...                 &#39;log2_Ratio M/L normalized BC36_2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; prot = pd.read_csv(&quot;_static/testdata/proteinGroups.zip&quot;, sep=&#39;\t&#39;, low_memory=False)</span>
<span class="sd">    &gt;&gt;&gt; protRatio = prot.filter(regex=&quot;Ratio .\/. normalized&quot;)</span>
<span class="sd">    &gt;&gt;&gt; protLog = pp.log(prot, protRatio, base=2)</span>
<span class="sd">    &gt;&gt;&gt; prot_tt = ana.ttest(df=protLog, reps=twitchVsmild, cond=&quot;TvM&quot;, mean=True, adjust_p_vals=False)</span>
<span class="sd">    &gt;&gt;&gt; prot_tt_adj = ana.adjust_p(prot_tt, p_col=&quot;pValue_TvM&quot;)</span>
<span class="sd">    &gt;&gt;&gt; prot_tt_adj.filter(regex=&#39;pValue&#39;).head()</span>
<span class="sd">       pValue_TvM  adj.pValue_TvM</span>
<span class="sd">    0         NaN             NaN</span>
<span class="sd">    1    0.947334        0.966514</span>
<span class="sd">    2         NaN             NaN</span>
<span class="sd">    3         NaN             NaN</span>
<span class="sd">    4    0.031292        0.206977</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># indices of rows containing values</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">p_col</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()]</span><span class="o">.</span><span class="n">index</span>
    <span class="c1"># init new col with for adjusted p-values</span>
    <span class="n">df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;adj.</span><span class="si">{</span><span class="n">p_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="c1"># apply correction for selected rows</span>
    <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;adj.</span><span class="si">{</span><span class="n">p_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">multipletests</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">p_col</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="cohen_d"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.cohen_d">[docs]</a><span class="k">def</span> <span class="nf">cohen_d</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group1</span><span class="p">,</span> <span class="n">group2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate Cohen&#39;s d effect size for two groups.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.Dataframe</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    group1 : str</span>
<span class="sd">        Colname for group1.</span>
<span class="sd">    group2 : str</span>
<span class="sd">        Colname for group2.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pd.Dataframe</span>
<span class="sd">        Input dataframe with a new column &quot;cohenD&quot;.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Cohen&#39;s d is defined as the difference between two means divided by a standard deviation for the data.</span>
<span class="sd">    Note that the pooled standard deviation here is calculated differently than</span>
<span class="sd">    originally proposed by Cohen.</span>

<span class="sd">    References</span>
<span class="sd">    ----------</span>
<span class="sd">    [1] Cohen, Jacob (1988). Statistical Power Analysis for the Behavioral Sciences. Routledge.</span>
<span class="sd">    [2] https://www.doi.org/10.22237/jmasm/1257035100</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mean1</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">std1</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">mean2</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">std2</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="c1"># TODO: the pooled sd here is calculated omitting the sample sizes n</span>
    <span class="c1"># This is not exactly what was proposed for cohens d: https://en.wikipedia.org/wiki/Effect_size</span>
    <span class="n">sd_pooled</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">((</span><span class="n">std1</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">std2</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cohenD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">mean1</span> <span class="o">-</span> <span class="n">mean2</span><span class="p">))</span> <span class="o">/</span> <span class="n">sd_pooled</span>
    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="AutoPCA"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA">[docs]</a><span class="k">class</span> <span class="nc">AutoPCA</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Conduct principal component analyses.</span>

<span class="sd">    The class encompasses a set of helpful visualizations</span>
<span class="sd">    for further investigating the results of the PCA</span>
<span class="sd">    It needs the matrix on which the PCA is performed</span>
<span class="sd">    as well as row labels (rlabels)</span>
<span class="sd">    and column labels (clabels) corresponding to the</span>
<span class="sd">    provided matrix.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    PCA is a method which allows you to visually investigate the underlying structure</span>
<span class="sd">    in your data after reduction of the dimensionality.</span>
<span class="sd">    With the .autoPCA() you can easily perform a PCA and also generate exploratory figures.</span>
<span class="sd">    Intro to PCA: https://learnche.org/pid/latent-variable-modelling/principal-component-analysis/index</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    for PCA no missing values are allowed</span>
<span class="sd">    filter those and store complete dataframe</span>

<span class="sd">    &gt;&gt;&gt; temp = prot[~prot.filter(regex=&quot;log2.*norm&quot;).isnull().any(1)]</span>

<span class="sd">    get the matrix of quantitative values corresponding to conditions of interest</span>
<span class="sd">    Here we only use the first replicate for clarity</span>

<span class="sd">    &gt;&gt;&gt; dataframe = temp.filter(regex=&quot;log2.*norm.*_1$&quot;)</span>

<span class="sd">    generate appropiate names for the columns and rows of the matrix</span>
<span class="sd">    for example here the columns represent the conditions and we are not interested in the rows (which are the genes)</span>

<span class="sd">    &gt;&gt;&gt; clabels = dataframe.columns</span>
<span class="sd">    &gt;&gt;&gt; rlabels = np.nan</span>

<span class="sd">    generate autopca object</span>

<span class="sd">    &gt;&gt;&gt; autopca = autoprot.analysis.AutoPCA(dataframe, rlabels, clabels)</span>

<span class="sd">    The scree plots describe how much of the total variance of the dataset is</span>
<span class="sd">    explained ba the first n components. As you want to explain as variance as</span>
<span class="sd">    possible with as little variables as possible, chosing the number of components</span>
<span class="sd">    directly right to the steep descend of the plot is usually a good idea.</span>

<span class="sd">    &gt;&gt;&gt; autopca.scree()</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import autoprot.preprocessing as pp</span>
<span class="sd">        import pandas as pd</span>

<span class="sd">        prot = pd.read_csv(&quot;_static/testdata/proteinGroups.zip&quot;, sep=&quot;\t&quot;, low_memory=False)</span>
<span class="sd">        protRatio = prot.filter(regex=&quot;Ratio .\/. normalized&quot;)</span>
<span class="sd">        protLog = pp.log(prot, protRatio, base=2)</span>
<span class="sd">        temp = protLog[~protLog.filter(regex=&quot;log2.*norm&quot;).isnull().any(1)]</span>
<span class="sd">        dataframe = temp.filter(regex=&quot;log2.*norm.*_1$&quot;)</span>
<span class="sd">        clabels = dataframe.columns</span>
<span class="sd">        rlabels = np.nan</span>
<span class="sd">        autopca = ana.AutoPCA(dataframe, rlabels, clabels)</span>
<span class="sd">        autopca.scree()</span>

<span class="sd">    The corrComp heatmap shows the PCA loads (i.e. how much a principal component is</span>
<span class="sd">    influenced by a change in that variable) relative to the variables (i.e. the</span>
<span class="sd">    experiment conditions). If a weight (colorbar) is close to zero, the corresponding</span>
<span class="sd">    PC is barely influenced by it.</span>

<span class="sd">    &gt;&gt;&gt; autopca.corr_comp(annot=False)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        autopca.corr_comp(annot=False)</span>

<span class="sd">    The bar loading plot is a different way to represent the weights/loads for each</span>
<span class="sd">    condition and principal component. High values indicate a high influence of the</span>
<span class="sd">    variable/condition on the PC.</span>

<span class="sd">    &gt;&gt;&gt; autopca.bar_load(pc=1)</span>
<span class="sd">    &gt;&gt;&gt; autopca.bar_load(pc=2)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        autopca.bar_load(pc=1)</span>
<span class="sd">        autopca.bar_load(pc=2)</span>

<span class="sd">    The score plot shows how the different data points (i.e. proteins) are positioned</span>
<span class="sd">    with respect to two principal components.</span>
<span class="sd">    In more detail, the scores are the original data values multiplied by the</span>
<span class="sd">    weights of each value for each principal component.</span>
<span class="sd">    Usually they will separate more in the direction of PC1 as this component</span>
<span class="sd">    explains the largest share of the data variance</span>

<span class="sd">    &gt;&gt;&gt; autopca.score_plot(pc1=1, pc2=2)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        autopca.score_plot(pc1=1, pc2=2)</span>

<span class="sd">    The loading plot is the 2D representation of the barLoading plots and shows</span>
<span class="sd">    the weights how each variable influences the two PCs.</span>

<span class="sd">    &gt;&gt;&gt; autopca.loading_plot(pc1=1, pc2=2, labeling=True)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        autopca.loading_plot(pc1=1, pc2=2, labeling=True)</span>

<span class="sd">    The Biplot is a combination of loading plot and score plot as it shows the</span>
<span class="sd">    scores for each protein as point and the weights for each variable as</span>
<span class="sd">    vectors.</span>
<span class="sd">    &gt;&gt;&gt; autopca.bi_plot(pc1=1, pc2=2)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        autopca.bi_plot(pc1=1, pc2=2)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># =========================================================================</span>
    <span class="c1"># TODO</span>
    <span class="c1"># - Add interactive 3D scatter plot</span>
    <span class="c1"># - Facilitate naming of columns and rows</span>
    <span class="c1"># - Allow further customization of plots (e.g. figsize)</span>
    <span class="c1"># - Implement pair plot for multiple dimensions</span>
    <span class="c1"># =========================================================================</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">rlabels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">clabels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">batch</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialise PCA class.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dataframe : pd.DataFrame</span>
<span class="sd">            Input dataframe.</span>
<span class="sd">        rlabels : list</span>
<span class="sd">            Row labels.</span>
<span class="sd">        clabels : list</span>
<span class="sd">            Column labels.</span>
<span class="sd">        batch : list, optional</span>
<span class="sd">            Labels for distinct conditions used to colour dots in score plot.</span>
<span class="sd">            Must be the length of rlabels.</span>
<span class="sd">            The default is None.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
            <span class="n">dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dataframe</span><span class="p">)</span>
        <span class="c1"># drop any rows in the dataframe containing missing values</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="n">clabels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rlabel</span> <span class="o">=</span> <span class="n">rlabels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">batch</span> <span class="o">=</span> <span class="n">batch</span>
        <span class="c1"># PCA is performed with the df containing missing values</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pca</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_perform_pca</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">clabels</span><span class="p">)</span>
        <span class="c1"># generate scores from loadings</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expVar</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_perform_pca</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">label</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Perform pca and generate for_vis dataframe.&quot;&quot;&quot;</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">dataframe</span><span class="o">.</span><span class="n">dropna</span><span class="p">())</span>
        <span class="c1"># components_ is and ndarray of shape (n_components, n_features)</span>
        <span class="c1"># and contains the loadings/weights of each PCA eigenvector</span>
        <span class="n">for_vis</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="n">for_vis</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">dataframe</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span>
        <span class="k">return</span> <span class="n">pca</span><span class="p">,</span> <span class="n">for_vis</span>

<div class="viewcode-block" id="AutoPCA.scree"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.scree">[docs]</a>    <span class="k">def</span> <span class="nf">scree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">)):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot Scree plot and Explained variance vs number of components.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        figsize : tuple of int, optional</span>
<span class="sd">            The size of the figure object.</span>
<span class="sd">            The default is (15,5).</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        TypeError</span>
<span class="sd">            No PCA object was initialised in the class.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pca</span><span class="p">,</span> <span class="n">PCA</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;This is a function to plot Scree plots. Provide fitted sklearn PCA object.&quot;</span><span class="p">)</span>

        <span class="n">eig_val</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_</span>
        <span class="n">cum_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">))</span>

        <span class="k">def</span> <span class="nf">_set_labels</span><span class="p">(</span><span class="n">ylabel</span><span class="p">,</span> <span class="n">title</span><span class="p">):</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;# Component&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">eig_val</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">eig_val</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;teal&quot;</span><span class="p">,</span>
                 <span class="n">markerfacecolor</span><span class="o">=</span><span class="s1">&#39;purple&#39;</span><span class="p">)</span>
        <span class="n">_set_labels</span><span class="p">(</span><span class="s2">&quot;Eigenvalues&quot;</span><span class="p">,</span> <span class="s2">&quot;Scree plot&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cum_var</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cum_var</span><span class="p">,</span> <span class="n">ds</span><span class="o">=</span><span class="s2">&quot;steps&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;teal&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">eig_val</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">_set_labels</span><span class="p">(</span><span class="s2">&quot;explained cumulative variance&quot;</span><span class="p">,</span> <span class="s2">&quot;Explained variance&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="AutoPCA.corr_comp"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.corr_comp">[docs]</a>    <span class="k">def</span> <span class="nf">corr_comp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot heatmap of PCA weights vs. variables.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        annot : bool, optional</span>
<span class="sd">            If True, write the data value in each cell.</span>
<span class="sd">            If an array-like with the same shape as data, then use this</span>
<span class="sd">            to annotate the heatmap instead of the data.</span>
<span class="sd">            Note that DataFrames will match on position, not index.</span>
<span class="sd">            The default is False.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        2D representation how strong each observation (e.g. log protein ratio)</span>
<span class="sd">        weights for each principal component.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;^PC&quot;</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;PuOr&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">annot</span><span class="o">=</span><span class="n">annot</span><span class="p">)</span>
        <span class="n">yp</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mf">0.5</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">))]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">yp</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="AutoPCA.bar_load"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.bar_load">[docs]</a>    <span class="k">def</span> <span class="nf">bar_load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">25</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot the loadings of a given component in a barplot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pc : int, optional</span>
<span class="sd">            Component to draw. The default is 1.</span>
<span class="sd">        n : int, optional</span>
<span class="sd">            Plot only the n first rows.</span>
<span class="sd">            The default is 25.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pc</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">for_vis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">for_vis</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">_abs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">for_vis</span><span class="p">[</span><span class="n">pc</span><span class="p">])</span>
        <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;negative&quot;</span>
        <span class="n">for_vis</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">for_vis</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;positive&quot;</span>
        <span class="n">for_vis</span> <span class="o">=</span> <span class="n">for_vis</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">_abs&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:</span><span class="n">n</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">()</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">for_vis</span><span class="p">[</span><span class="n">pc</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">],</span> <span class="n">hue</span><span class="o">=</span><span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span>
                    <span class="n">hue_order</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;negative&quot;</span><span class="p">,</span> <span class="s2">&quot;positive&quot;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;teal&quot;</span><span class="p">,</span> <span class="s2">&quot;purple&quot;</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">get_legend</span><span class="p">()</span><span class="o">.</span><span class="n">remove</span><span class="p">()</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span></div>

<div class="viewcode-block" id="AutoPCA.return_load"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.return_load">[docs]</a>    <span class="k">def</span> <span class="nf">return_load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">25</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the load for a given principal component.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pc : int, optional</span>
<span class="sd">            Component to draw. The default is 1.</span>
<span class="sd">        n : int, optional</span>
<span class="sd">            Plot only the n first rows.</span>
<span class="sd">            The default is 25.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            Dataframe containing load vs. condition.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pc</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">for_vis</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">for_vis</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">_abs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">for_vis</span><span class="p">[</span><span class="n">pc</span><span class="p">])</span>
        <span class="n">for_vis</span> <span class="o">=</span> <span class="n">for_vis</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pc</span><span class="si">}</span><span class="s2">_abs&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:</span><span class="n">n</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">for_vis</span><span class="p">[[</span><span class="n">pc</span><span class="p">,</span> <span class="s2">&quot;label&quot;</span><span class="p">]]</span></div>

<div class="viewcode-block" id="AutoPCA.return_score"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.return_score">[docs]</a>    <span class="k">def</span> <span class="nf">return_score</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a dataframe of all scorings for all principal components.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        scores : pd.DataFrame</span>
<span class="sd">            Dataframe holding the principal components as colnames and</span>
<span class="sd">            the scores for each protein on that PC as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">scores</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span>
        <span class="k">return</span> <span class="n">scores</span></div>

<div class="viewcode-block" id="AutoPCA.score_plot"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.score_plot">[docs]</a>    <span class="k">def</span> <span class="nf">score_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc1</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pc2</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">labeling</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">)):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a PCA score plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pc1 : int, optional</span>
<span class="sd">            Number of the first PC to plot. The default is 1.</span>
<span class="sd">        pc2 : int, optional</span>
<span class="sd">            Number of the second PC to plot. The default is 2.</span>
<span class="sd">        labeling : bool, optional</span>
<span class="sd">            If True, points are labelled with the corresponding</span>
<span class="sd">            column labels. The default is False.</span>
<span class="sd">        file : str, optional</span>
<span class="sd">            Path to save the plot. The default is None.</span>
<span class="sd">        figsize : tuple of int, optional</span>
<span class="sd">            Figure size. The default is (5,5).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This will return a scatterplot with as many points as there are</span>
<span class="sd">        entries (i.e. protein IDs).</span>
<span class="sd">        The scores for each PC are the original protein ratios multiplied with</span>
<span class="sd">        the loading weights.</span>
<span class="sd">        The score plot corresponds to the individual positions of of each protein</span>
<span class="sd">        on a hyperplane generated by the pc1 and pc2 vectors.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">for_vis</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">y</span><span class="p">})</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">for_vis</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">for_vis</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">y</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span><span class="p">})</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">for_vis</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">])</span>
        <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rlabel</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Score plot&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">labeling</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">ss</span> <span class="o">=</span> <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span>
            <span class="n">xx</span> <span class="o">=</span> <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span>
            <span class="n">yy</span> <span class="o">=</span> <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span> <span class="n">yy</span><span class="p">,</span> <span class="n">ss</span><span class="p">):</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">fr</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">/ScorePlot.pdf&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="AutoPCA.loading_plot"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.loading_plot">[docs]</a>    <span class="k">def</span> <span class="nf">loading_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc1</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pc2</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">labeling</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">)):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a PCA loading plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pc1 : int, optional</span>
<span class="sd">            Number of the first PC to plot. The default is 1.</span>
<span class="sd">        pc2 : int, optional</span>
<span class="sd">            Number of the second PC to plot. The default is 2.</span>
<span class="sd">        labeling : bool, optional</span>
<span class="sd">            If True, points are labelled with the corresponding</span>
<span class="sd">            column labels. The default is False.</span>
<span class="sd">        figsize : tuple of int, optional</span>
<span class="sd">            The size of the figure object.</span>
<span class="sd">            The default is (5,5).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This will return a scatterplot with as many points as there are</span>
<span class="sd">        components (i.e. conditions) in the dataset.</span>
<span class="sd">        For each component a load magnitude for two PCs will be printedd</span>
<span class="sd">        that describes how much each condition influences the magnitude</span>
<span class="sd">        of the respective PC.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">batch</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">batch</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Loadings plot&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">labeling</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">ss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span>
            <span class="n">xx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="n">yy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">xx</span><span class="p">,</span> <span class="n">yy</span><span class="p">,</span> <span class="n">ss</span><span class="p">):</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span></div>

<div class="viewcode-block" id="AutoPCA.bi_plot"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.bi_plot">[docs]</a>    <span class="k">def</span> <span class="nf">bi_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc1</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pc2</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">num_load</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate a biplot, a combined loadings and score plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        pc1 : int, optional</span>
<span class="sd">            Number of the first PC to plot. The default is 1.</span>
<span class="sd">        pc2 : int, optional</span>
<span class="sd">            Number of the second PC to plot. The default is 2.</span>
<span class="sd">        num_load : &#39;all&#39; or int, optional</span>
<span class="sd">            Plot only the n first rows.</span>
<span class="sd">            The default is &quot;all&quot;.</span>
<span class="sd">        figsize : tuple of int, optional</span>
<span class="sd">            Figure size. The default is (3,3).</span>
<span class="sd">        **kwargs :</span>
<span class="sd">            Passed to plt.scatter.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        In the biplot, scores are shown as points and loadings as</span>
<span class="sd">        vectors.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;lightgray&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">temp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">forVis</span><span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]]</span>
        <span class="n">temp</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">num_load</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
            <span class="n">loadings</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loadings</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">num_load</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">num_load</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="n">xscale</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">yscale</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[::,</span> <span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">xmina</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">xmaxa</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ymina</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">load</span><span class="p">,</span> <span class="n">lab</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">labels</span><span class="p">):</span>
            <span class="c1"># plt.plot([0,load[0]/xscale], (0, load[1]/yscale), color=&quot;purple&quot;)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">arrow</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dx</span><span class="o">=</span><span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span><span class="p">,</span> <span class="n">dy</span><span class="o">=</span><span class="n">load</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">yscale</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;purple&quot;</span><span class="p">,</span>
                      <span class="n">head_width</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">load</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">yscale</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">lab</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span> <span class="o">&lt;</span> <span class="n">xmina</span><span class="p">:</span>
                <span class="n">xmina</span> <span class="o">=</span> <span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span>
            <span class="k">elif</span> <span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span> <span class="o">&gt;</span> <span class="n">xmaxa</span><span class="p">:</span>
                <span class="n">xmaxa</span> <span class="o">=</span> <span class="n">load</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">xscale</span>

            <span class="k">if</span> <span class="n">load</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">yscale</span> <span class="o">&lt;</span> <span class="n">ymina</span> <span class="ow">or</span> <span class="n">load</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">yscale</span> <span class="o">&gt;</span> <span class="n">ymina</span><span class="p">:</span>
                <span class="n">ymina</span> <span class="o">=</span> <span class="n">load</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">yscale</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc1</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc1</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">pc2</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expVar</span><span class="p">[</span><span class="n">pc2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2"> %&quot;</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span></div>

<div class="viewcode-block" id="AutoPCA.pair_plot"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.AutoPCA.pair_plot">[docs]</a>    <span class="k">def</span> <span class="nf">pair_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Draw a pair plot of for pca for the given number of dimensions.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        n : int, optional</span>
<span class="sd">            Plot only the n first rows. The default is 0.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Be careful for large data this might crash you PC -&gt; better specify n!</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO must be prettyfied quite a bit....</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">for_vis</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="p">[:,</span> <span class="p">:</span><span class="n">n</span><span class="p">])</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="n">pcs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Xt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">for_vis</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">pcs</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">for_vis</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span><span class="n">for_vis</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span><span class="n">for_vis</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="KSEA"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA">[docs]</a><span class="k">class</span> <span class="nc">KSEA</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform kinase substrate enrichment analysis.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    KSEA uses the Kinase-substrate dataset and the</span>
<span class="sd">    regulatory-sites dataset from https://www.phosphosite.org/staticDownloads</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    KSEA is a method to get insights on which kinases are active in a given</span>
<span class="sd">    phosphoproteomic dataset. This is a great method to gain deeper insights</span>
<span class="sd">    on the underlying signaling mechanisms and also to generate novel</span>
<span class="sd">    hypothesis and find new connections in signaling processes.</span>
<span class="sd">    The KSEA class allows you to easily perform the analysis and</span>
<span class="sd">    comes with helpful functions to visualize and interpret your results.</span>

<span class="sd">    In the first step of the analysis you have to generate a KSEA object.</span>

<span class="sd">    &gt;&gt;&gt; ksea = autoprot.analysis.KSEA(phos)</span>

<span class="sd">    Next, you can annotate the data with respective kinases.</span>
<span class="sd">    You can provide the function with a organism of your choice as well as</span>
<span class="sd">    toggle whether or not to screen for only in vivo determined substrate</span>
<span class="sd">    phosphorylation of the respective kinases.</span>

<span class="sd">    &gt;&gt;&gt; ksea.annotate(organism=&quot;mouse&quot;, only_in_vivo=True)</span>

<span class="sd">    After the annotation it is always a good idea to get an overview of the</span>
<span class="sd">    kinases in the data an how many substrates the have. Based on this you</span>
<span class="sd">    might want to adjust a cutoff specifying the minimum number of substrates</span>
<span class="sd">    per kinase.</span>

<span class="sd">    &gt;&gt;&gt; ksea.get_kinase_overview(kois=[&quot;Akt1&quot;,&quot;MKK4&quot;, &quot;P38A&quot;, &quot;Erk1&quot;])</span>

<span class="sd">    Next, you can perform the actual kinase substrate enrichment analysis.</span>
<span class="sd">    The analysis is based on the log fold change of your data.</span>
<span class="sd">    Therefore, you have to provide the function with the appropiate column of</span>
<span class="sd">    your data and the minimum number of substrates per kinase.</span>

<span class="sd">    &gt;&gt;&gt; ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5)</span>

<span class="sd">    After the ksea has finished, you can get information for further analysis</span>
<span class="sd">    such as the substrates of a specific kinase (or a list of kinases)</span>

<span class="sd">    &gt;&gt;&gt; ksea.return_kinase_substrate(kinase=[&quot;Akt1&quot;, &quot;MKK4&quot;]).sample() # doctest: +SKIP</span>

<span class="sd">    or a new dataframe with additional columns for every kinase showing if the</span>
<span class="sd">    protein is a substrate of that kinase or not</span>

<span class="sd">    &gt;&gt;&gt; ksea.annotate_df(kinases=[&quot;Akt1&quot;, &quot;MKK4&quot;]).iloc[:2,-5:]</span>

<span class="sd">    Eventually, you can also generate plots of the enrichment analysis.</span>

<span class="sd">    &gt;&gt;&gt; ksea.plot_enrichment(up_col=&quot;salmon&quot;)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.preprocessing as pp</span>
<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import pandas as pd</span>

<span class="sd">        phos = pd.read_csv(&quot;_static/testdata/Phospho (STY)Sites_mod.zip&quot;, sep=&quot;\t&quot;, low_memory=False)</span>
<span class="sd">        phos = pp.cleaning(phos, file = &quot;Phospho (STY)&quot;)</span>
<span class="sd">        phosRatio = phos.filter(regex=&quot;^Ratio .\/.( | normalized )R.___&quot;).columns</span>
<span class="sd">        phos = pp.log(phos, phosRatio, base=2)</span>
<span class="sd">        phos = pp.filter_loc_prob(phos, thresh=.75)</span>
<span class="sd">        phosRatio = phos.filter(regex=&quot;log2_Ratio .\/.( | normalized )R.___&quot;).columns</span>
<span class="sd">        phos = pp.remove_non_quant(phos, phosRatio)</span>

<span class="sd">        phosRatio = phos.filter(regex=&quot;log2_Ratio .\/. normalized R.___&quot;).columns</span>
<span class="sd">        phos_expanded = pp.expand_site_table(phos, phosRatio)</span>

<span class="sd">        twitchVsmild = [&#39;log2_Ratio H/M normalized R1&#39;,&#39;log2_Ratio M/L normalized R2&#39;,&#39;log2_Ratio H/M normalized R3&#39;,</span>
<span class="sd">                        &#39;log2_Ratio H/L normalized R4&#39;,&#39;log2_Ratio H/M normalized R5&#39;,&#39;log2_Ratio M/L normalized R6&#39;]</span>
<span class="sd">        twitchVsctrl = [&quot;log2_Ratio H/L normalized R1&quot;,&quot;log2_Ratio H/M normalized R2&quot;,&quot;log2_Ratio H/L normalized R3&quot;,</span>
<span class="sd">                        &quot;log2_Ratio M/L normalized R4&quot;, &quot;log2_Ratio H/L normalized R5&quot;,&quot;log2_Ratio H/M normalized R6&quot;]</span>

<span class="sd">        phos = ana.ttest(df=phos_expanded, reps=twitchVsmild, cond=&quot;_TvM&quot;)</span>
<span class="sd">        phos = ana.ttest(df=phos_expanded, reps=twitchVsctrl, cond=&quot;_TvC&quot;)</span>

<span class="sd">        ksea = ana.KSEA(phos)</span>
<span class="sd">        ksea.annotate(organism=&quot;mouse&quot;, only_in_vivo=True)</span>
<span class="sd">        ksea.get_kinase_overview(kois=[&quot;Akt1&quot;,&quot;MKK4&quot;, &quot;P38A&quot;, &quot;Erk1&quot;])</span>
<span class="sd">        ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5)</span>

<span class="sd">        ksea.plot_enrichment(up_col=&quot;salmon&quot;)</span>

<span class="sd">    You can also highlight a list of kinases in volcano plots.</span>
<span class="sd">    This is based on the autoprot volcano function.</span>
<span class="sd">    You can pass all the common parameters to this function.</span>

<span class="sd">    &gt;&gt;&gt; ksea.volcanos(log_fc=&quot;logFC_TvC&quot;, p=&quot;pValue_TvC&quot;, kinases=[&quot;Akt1&quot;, &quot;MKK4&quot;],</span>
<span class="sd">    ...               annot=&quot;Gene names&quot;, sig_col=&quot;gray&quot;)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        ksea.volcanos(log_fc=&quot;logFC_TvC&quot;, p=&quot;pValue_TvC&quot;, kinases=[&quot;Akt1&quot;, &quot;MKK4&quot;],</span>
<span class="sd">                      annot=&quot;Gene names&quot;, sig_col=&quot;gray&quot;)</span>

<span class="sd">    Sometimes the enrichment is crowded by various kinase isoforms.</span>
<span class="sd">    In such cases it makes sense to simplify the annotation by grouping those</span>
<span class="sd">    isoforms together.</span>

<span class="sd">    &gt;&gt;&gt; simplify = {&quot;ERK&quot;:[&quot;ERK1&quot;,&quot;ERK2&quot;],</span>
<span class="sd">    ...             &quot;GSK3&quot;:[&quot;GSK3A&quot;, &quot;GSK3B&quot;]}</span>
<span class="sd">    &gt;&gt;&gt; ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5, simplify=simplify)</span>
<span class="sd">    &gt;&gt;&gt; ksea.plot_enrichment()</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        simplify = {&quot;ERK&quot;:[&quot;ERK1&quot;,&quot;ERK2&quot;],</span>
<span class="sd">                    &quot;GSK3&quot;:[&quot;GSK3A&quot;, &quot;GSK3B&quot;]}</span>
<span class="sd">        ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5, simplify=simplify)</span>
<span class="sd">        ksea.plot_enrichment()</span>

<span class="sd">    Of course, you can also get the ksea results as a dataframe to save or to further customize.</span>

<span class="sd">    &gt;&gt;&gt; ksea.return_enrichment()</span>

<span class="sd">    Of course is the database not exhaustive and you might want to add additional</span>
<span class="sd">    substrates manually. This can be done the following way.</span>
<span class="sd">    Manually added substrates are always added irrespective of the species used</span>
<span class="sd">    for the annotation.</span>

<span class="sd">    &gt;&gt;&gt; ksea = ana.KSEA(phos)</span>
<span class="sd">    &gt;&gt;&gt; genes = [&quot;RPGR&quot;]</span>
<span class="sd">    &gt;&gt;&gt; modRsds = [&quot;S564&quot;]</span>
<span class="sd">    &gt;&gt;&gt; kinases = [&quot;mTOR&quot;]</span>
<span class="sd">    &gt;&gt;&gt; ksea.add_substrate(kinase=kinases, substrate=genes, sub_mod_rsd=modRsds)</span>

<span class="sd">    &gt;&gt;&gt; ksea.annotate(organism=&quot;mouse&quot;, only_in_vivo=True)</span>
<span class="sd">    &gt;&gt;&gt; ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5)</span>
<span class="sd">    &gt;&gt;&gt; ksea.plot_enrichment(plot_bg=False)</span>

<span class="sd">    &gt;&gt;&gt; ksea.clear_manual_substrates()</span>
<span class="sd">    &gt;&gt;&gt; ksea.annotate(organism=&quot;mouse&quot;, only_in_vivo=True)</span>
<span class="sd">    &gt;&gt;&gt; ksea.ksea(col=&quot;logFC_TvC&quot;, min_subs=5)</span>
<span class="sd">    &gt;&gt;&gt; ksea.plot_enrichment(plot_bg=False)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialise the KSEA object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data : pd.DataFrame</span>
<span class="sd">            matrix_a phosphoproteomics datasaet.</span>
<span class="sd">            This data has to contain information about Gene name, position and amino acid of the peptides with</span>
<span class="sd">            &quot;Gene names&quot;, &quot;Position&quot; and &quot;Amino acid&quot; as the respective column names.</span>
<span class="sd">            Optionally you can provide a &quot;Multiplicity&quot; column.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">resources</span><span class="o">.</span><span class="n">open_binary</span><span class="p">(</span><span class="s2">&quot;autoprot.data&quot;</span><span class="p">,</span> <span class="s2">&quot;Kinase_Substrate_Dataset.zip&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">d</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;zip&#39;</span><span class="p">)</span>
        <span class="c1"># harmonize gene naming</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;SUB_GENE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;SUB_GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="c1"># add source information</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;PSP&quot;</span>
        <span class="k">with</span> <span class="n">resources</span><span class="o">.</span><span class="n">open_binary</span><span class="p">(</span><span class="s2">&quot;autoprot.data&quot;</span><span class="p">,</span> <span class="s2">&quot;Regulatory_sites.zip&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">d</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">PSP_regSits</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;zip&#39;</span><span class="p">)</span>
        <span class="c1"># Harmonize the input data and store them to the class</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_preprocess</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
        <span class="c1"># init other class objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">koi</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">simpleDf</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_preprocess</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Define MOD_RSD, ucGene and mergeID cols in the input dataset.&quot;&quot;&quot;</span>
        <span class="c1"># New column containing the modified residue as Ser201</span>
        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;MOD_RSD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Amino acid&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="c1"># The Gene names as defined for the Kinase substrate dataset</span>
        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Gene names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="c1"># an index column</span>
        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;mergeID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">data</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_enrichment</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">kinase</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the enrichment score for a certain kinase.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        df : pd.Dataframe</span>
<span class="sd">            Input datafame with enrichment information.</span>
<span class="sd">        col : str</span>
<span class="sd">            Column containing enrichment information e.g. intensity ratios.</span>
<span class="sd">            Must be present in df.</span>
<span class="sd">        kinase : str</span>
<span class="sd">            Kinase to calculate the enrichment for.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list</span>
<span class="sd">            pair of kinase name and score.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get enrichment values for rows containing the kinase of interest</span>
        <span class="n">ks</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">][</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;KINASE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">x</span><span class="p">)]</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">ks</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>  <span class="c1"># mean FC of kinase subs</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>  <span class="c1"># mean FC of all substrates</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">ks</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># number of kinase substrates</span>
        <span class="n">sig</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>  <span class="c1"># standard dev of FC of all</span>
        <span class="n">score</span> <span class="o">=</span> <span class="p">((</span><span class="n">s</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">m</span><span class="p">))</span> <span class="o">/</span> <span class="n">sig</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="n">score</span><span class="p">]</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_extract_kois</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Count the number of substrates for each kinase in a merged df.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        df : pd.DataFrame</span>
<span class="sd">            Merged dataframe containing kinase substrate pairs present in the</span>
<span class="sd">            input dataframe.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            Dataframe with columns &quot;Kinase&quot; and &quot;#Subs&quot; containing the</span>
<span class="sd">            numbers of appearances of each kinase in the merged input dataset.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Extract all strings present in the KINASE column as list of str</span>
        <span class="c1"># This is mainly out of caution as all entries in the kinase col should be</span>
        <span class="c1"># strings</span>
        <span class="n">koi</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;KINASE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">)]</span>
        <span class="c1"># remove duplicates</span>
        <span class="n">ks</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">koi</span><span class="p">)</span>
        <span class="c1"># empty list to take on sets of kinase:count pairs</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="p">[(</span><span class="n">k</span><span class="p">,</span> <span class="n">koi</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">k</span><span class="p">))</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ks</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Kinase&quot;</span><span class="p">,</span> <span class="s2">&quot;#Subs&quot;</span><span class="p">])</span>

<div class="viewcode-block" id="KSEA.add_substrate"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.add_substrate">[docs]</a>    <span class="k">def</span> <span class="nf">add_substrate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinase</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">substrate</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">sub_mod_rsd</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Manually add a substrate to the database.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinase : list of str</span>
<span class="sd">            Name of the kinase e.g. PAK2.</span>
<span class="sd">        substrate : list of str</span>
<span class="sd">            Name of the substrate e.g. Prkd1.</span>
<span class="sd">        sub_mod_rsd : list of str</span>
<span class="sd">            Phosphorylated residues e.g. S203.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the three provided lists do not match in length.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># a bit cumbersome way to check if all lists</span>
        <span class="c1"># are of the same lengths</span>
        <span class="n">it</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">([</span><span class="n">kinase</span><span class="p">,</span> <span class="n">substrate</span><span class="p">,</span> <span class="n">sub_mod_rsd</span><span class="p">])</span>
        <span class="n">the_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">it</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">!=</span> <span class="n">the_len</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">it</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;not all lists have same length!&#39;</span><span class="p">)</span>

        <span class="c1"># generate new empty df to fill in the new kinases</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">kinase</span><span class="p">)):</span>
            <span class="n">temp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;KINASE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kinase</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">temp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;SUB_GENE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">substrate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">temp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;SUB_MOD_RSD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sub_mod_rsd</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">temp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;manual&quot;</span>
        <span class="c1"># append to the original database from PSP</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="KSEA.clear_manual_substrates"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.clear_manual_substrates">[docs]</a>    <span class="k">def</span> <span class="nf">clear_manual_substrates</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Remove all manual entries from the PSP database.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;PSP&quot;</span><span class="p">]</span></div>

<div class="viewcode-block" id="KSEA.annotate"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.annotate">[docs]</a>    <span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">only_in_vivo</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Annotate with known kinase substrate pairs.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        organism : str, optional</span>
<span class="sd">            The target organism. The default is &quot;human&quot;.</span>
<span class="sd">        only_in_vivo : bool, optional</span>
<span class="sd">            Whether to restrict analysis to in vivo evidence.</span>
<span class="sd">            The default is False.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Manually added kinases will be included in the annotation search</span>
<span class="sd">        independent of the setting of organism and onInVivo.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># return a kinase substrate dataframe including only entries of the</span>
        <span class="c1"># target organism that were validated in vitro</span>
        <span class="k">if</span> <span class="n">only_in_vivo</span><span class="p">:</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;KIN_ORGANISM&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">organism</span><span class="p">)</span> <span class="o">&amp;</span>
                                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;SUB_ORGANISM&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">organism</span><span class="p">)</span> <span class="o">&amp;</span>
                                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;IN_VIVO_RXN&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;X&quot;</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;manual&quot;</span><span class="p">)]</span>
        <span class="c1"># only filter for the target organism</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;KIN_ORGANISM&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">organism</span><span class="p">)</span> <span class="o">&amp;</span>
                                <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;SUB_ORGANISM&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">organism</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PSP_KS</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;manual&quot;</span><span class="p">)]</span>

        <span class="c1"># merge the kinase substrate data tables with the input dataframe</span>
        <span class="c1"># include the multiplicity column in the merge if present in the</span>
        <span class="c1"># input dataframe</span>
        <span class="c1"># the substrate gene names and the modification position are used for</span>
        <span class="c1"># merging</span>
        <span class="k">if</span> <span class="s2">&quot;Multiplicity&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">,</span> <span class="s2">&quot;Multiplicity&quot;</span><span class="p">,</span> <span class="s2">&quot;mergeID&quot;</span><span class="p">]],</span>
                                    <span class="n">temp</span><span class="p">,</span>
                                    <span class="n">left_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">],</span>
                                    <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SUB_GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;SUB_MOD_RSD&quot;</span><span class="p">],</span>
                                    <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>  <span class="c1"># keep only entries that are present in the input dataframe</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">,</span> <span class="s2">&quot;mergeID&quot;</span><span class="p">]],</span>
                                    <span class="n">temp</span><span class="p">,</span>
                                    <span class="n">left_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">],</span>
                                    <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SUB_GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;SUB_MOD_RSD&quot;</span><span class="p">],</span>
                                    <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

        <span class="c1"># generate a df with kinase:number of substrate pairs for the dataset</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">koi</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_extract_kois</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span><span class="p">)</span></div>

    <span class="c1"># noinspection PyBroadException</span>
<div class="viewcode-block" id="KSEA.get_kinase_overview"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.get_kinase_overview">[docs]</a>    <span class="k">def</span> <span class="nf">get_kinase_overview</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kois</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot a graphical overview of the kinases acting on the proteins in the dataset.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kois : list of str, optional</span>
<span class="sd">            Kinases of interest for which a detailed overview of substrate numbers</span>
<span class="sd">            is plotted. The default is None.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># ax[0] is a histogram of kinase substrate numbers and</span>
        <span class="c1"># ax[1] is a table of top10 kinases</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="s2">&quot;#Subs&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Overview of #Subs per kinase&quot;</span><span class="p">)</span>

        <span class="c1"># get axis[1] ready - basically remove everthing</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;left&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;bottom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>  <span class="c1"># changes apply to the x-axis</span>
                          <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>  <span class="c1"># both major and minor ticks are affected</span>
                          <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># ticks along the bottom edge are off</span>
                          <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># ticks along the top edge are off</span>
                          <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># labels along the bottom edge are off</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

        <span class="c1"># plot table</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="s2">&quot;Top10</span><span class="se">\n</span><span class="s2">Kinase&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="mf">0.01</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="s2">&quot;#Subs&quot;</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">],</span> <span class="p">[</span><span class="mf">.975</span><span class="p">,</span> <span class="mf">.975</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">.975</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>

        <span class="c1"># get top 10 kinases for annotation</span>
        <span class="n">text</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;#Subs&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">text</span><span class="p">):</span>
            <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="n">j</span> <span class="o">/</span> <span class="mi">10</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mf">0.125</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="n">j</span> <span class="o">/</span> <span class="mi">10</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># plot some descriptive stats</span>
        <span class="n">tot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">s</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Substrates for </span><span class="si">{</span><span class="n">tot</span><span class="si">}</span><span class="s2"> kinases found in data.&quot;</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.975</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="n">med</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="s1">&#39;#Subs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Median #Sub: </span><span class="si">{</span><span class="n">med</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="n">mea</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="s1">&#39;#Subs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Mean #Sub: </span><span class="si">{</span><span class="n">mea</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.875</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="c1"># if kois are provided plot those</span>
        <span class="k">if</span> <span class="n">kois</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="mf">.8</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">kois</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="s2">&quot;Kinase&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="o">==</span> <span class="n">k</span><span class="o">.</span><span class="n">upper</span><span class="p">()][</span><span class="s2">&quot;#Subs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                    <span class="n">s</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">ss</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> has </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> substrates.&quot;</span>
                <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ss</span><span class="p">)</span>
                <span class="n">pos</span> <span class="o">-=</span> <span class="mf">0.055</span></div>

<div class="viewcode-block" id="KSEA.ksea"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.ksea">[docs]</a>    <span class="k">def</span> <span class="nf">ksea</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">min_subs</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">simplify</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Kinase Enrichment Score.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        col : str</span>
<span class="sd">            Column used for the analysis containing the kinase substrate</span>
<span class="sd">            enrichments.</span>
<span class="sd">        min_subs : int, optional</span>
<span class="sd">            Minimum number of substrates a kinase must have to be considered.</span>
<span class="sd">            The default is 5.</span>
<span class="sd">        simplify : None, &quot;auto&quot; or dict, optional</span>
<span class="sd">            Merge multiple kinases during analysis.</span>
<span class="sd">            Using &quot;auto&quot; a predefined set of kinase isoforms is merged.</span>
<span class="sd">            If provided with a dict, the dict has to contain a list of kinases</span>
<span class="sd">            to merge as values and the name of the merged kinases as key.</span>
<span class="sd">            The default is None.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The enrichment score is calculated as</span>

<span class="sd">        .. math::</span>
<span class="sd">            \frac{(\langle FC_{kinase} \rangle - \langle FC_{all} \rangle)\sqrt{N_{all}}}{\sigma_{all}}</span>

<span class="sd">        i.e. the difference in mean fold change between kinase and all substrates</span>
<span class="sd">        multiplied by the square root of number of kinase substrates and divided</span>
<span class="sd">        by the standard deviation of the fold change of all substrates.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO wouldn&#39;t it make more sense to perform simplification in the</span>
        <span class="c1"># Annotate function?</span>
        <span class="n">copy_annot_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">simplify</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">simplify</span> <span class="o">==</span> <span class="s2">&quot;auto&quot;</span><span class="p">:</span>
                <span class="n">simplify</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;AKT&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Akt1&quot;</span><span class="p">,</span> <span class="s2">&quot;Akt2&quot;</span><span class="p">,</span> <span class="s2">&quot;Akt3&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;PKC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;PKCA&quot;</span><span class="p">,</span> <span class="s2">&quot;PKCD&quot;</span><span class="p">,</span> <span class="s2">&quot;PKCE&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;ERK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ERK1&quot;</span><span class="p">,</span> <span class="s2">&quot;ERK2&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;GSK3&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;GSK3B&quot;</span><span class="p">,</span> <span class="s2">&quot;GSK3A&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;JNK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;JNK1&quot;</span><span class="p">,</span> <span class="s2">&quot;JNK2&quot;</span><span class="p">,</span> <span class="s2">&quot;JNK3&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;FAK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;FAK iso2&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;p70S6K&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;p70S6K&quot;</span><span class="p">,</span> <span class="s2">&quot;p70SKB&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;RSK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;p90RSK&quot;</span><span class="p">,</span> <span class="s2">&quot;RSK2&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;P38&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;P38A&quot;</span><span class="p">,</span> <span class="s2">&quot;P38B&quot;</span><span class="p">,</span> <span class="s2">&quot;P38C&quot;</span><span class="p">,</span> <span class="s2">&quot;P338D&quot;</span><span class="p">]}</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">simplify</span><span class="p">:</span>
                <span class="n">copy_annot_df</span><span class="p">[</span><span class="s2">&quot;KINASE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">simplify</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">simplify</span><span class="p">[</span><span class="n">key</span><span class="p">]),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># drop rows which are now duplicates</span>
            <span class="k">if</span> <span class="s2">&quot;Multiplicity&quot;</span> <span class="ow">in</span> <span class="n">copy_annot_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">copy_annot_df</span><span class="p">[[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">,</span> <span class="s2">&quot;Multiplicity&quot;</span><span class="p">,</span> <span class="s2">&quot;KINASE&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">index</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">idx</span> <span class="o">=</span> <span class="n">copy_annot_df</span><span class="p">[[</span><span class="s2">&quot;ucGene&quot;</span><span class="p">,</span> <span class="s2">&quot;MOD_RSD&quot;</span><span class="p">,</span> <span class="s2">&quot;KINASE&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">index</span>
            <span class="n">copy_annot_df</span> <span class="o">=</span> <span class="n">copy_annot_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">simpleDf</span> <span class="o">=</span> <span class="n">copy_annot_df</span>

            <span class="c1"># repeat annotation with the simplified dataset</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">koi</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_extract_kois</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">simpleDf</span><span class="p">)</span>

        <span class="c1"># filter kinases with at least min_subs number of substrates</span>
        <span class="n">koi</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">koi</span><span class="p">[</span><span class="s2">&quot;#Subs&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_subs</span><span class="p">][</span><span class="s2">&quot;Kinase&quot;</span><span class="p">]</span>

        <span class="c1"># init empty df</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;kinase&quot;</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">])</span>
        <span class="c1"># add the enrichment column back to the annotation df using the mergeID</span>
        <span class="n">copy_annot_df</span> <span class="o">=</span> <span class="n">copy_annot_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[[</span><span class="n">col</span><span class="p">,</span> <span class="s2">&quot;mergeID&quot;</span><span class="p">]],</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;mergeID&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">koi</span><span class="p">:</span>
            <span class="c1"># calculate the enrichment score</span>
            <span class="n">k</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_enrichment</span><span class="p">(</span><span class="n">copy_annot_df</span><span class="p">[</span><span class="n">copy_annot_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()],</span> <span class="n">col</span><span class="p">,</span> <span class="n">kinase</span><span class="p">)</span>
            <span class="c1"># new dataframe containing kinase names and scores</span>
            <span class="n">temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;kinase&quot;</span><span class="p">:</span> <span class="n">k</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="n">s</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="c1"># add the new df to the pre-initialised df</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># sort the concatenated dfs by kinase enrichment score</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>

<div class="viewcode-block" id="KSEA.return_enrichment"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.return_enrichment">[docs]</a>    <span class="k">def</span> <span class="nf">return_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a dataframe of kinase:score pairs.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First perform the enrichment&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># dropna in case of multiple columns in data</span>
            <span class="c1"># sometimes there are otherwise nan</span>
            <span class="c1"># nans are dropped in ksea enrichment</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span></div>

<div class="viewcode-block" id="KSEA.plot_enrichment"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.plot_enrichment">[docs]</a>    <span class="k">def</span> <span class="nf">plot_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">up_col</span><span class="o">=</span><span class="s2">&quot;orange&quot;</span><span class="p">,</span> <span class="n">down_col</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">bg_col</span><span class="o">=</span><span class="s2">&quot;lightgray&quot;</span><span class="p">,</span>
                        <span class="n">plot_bg</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ret</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">)):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot the KSEA results.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        up_col : str, optional</span>
<span class="sd">            Color for enriched/upregulated kinases.</span>
<span class="sd">            The default is &quot;orange&quot;.</span>
<span class="sd">        down_col : str, optional</span>
<span class="sd">            Colour for deriched/downregulated kinases.</span>
<span class="sd">            The default is &quot;blue&quot;.</span>
<span class="sd">        bg_col : str, optional</span>
<span class="sd">            Colour for not kinases that did not change significantly.</span>
<span class="sd">            The default is &quot;lightgray&quot;.</span>
<span class="sd">        plot_bg : bool, optional</span>
<span class="sd">            Whether to plot the unaffected kinases.</span>
<span class="sd">            The default is True.</span>
<span class="sd">        ret : bool, optional</span>
<span class="sd">            Whether to return the figure object.</span>
<span class="sd">            The default is False.</span>
<span class="sd">        title : str, optional</span>
<span class="sd">            Title of the figure. The default is &quot;&quot;.</span>
<span class="sd">        figsize : tuple of int, optional</span>
<span class="sd">            Figure size. The default is (5,10).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        fig : matplotlib figure.</span>
<span class="sd">            Only returned in ret is True.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First perform the enrichment&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># set all proteins to bg_col</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bg_col</span>
            <span class="c1"># highlight up and down regulated</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">up_col</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">down_col</span>
            <span class="c1"># init figure</span>
            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
            <span class="c1"># only plot the unaffected substrates if plot_bg is True</span>
            <span class="k">if</span> <span class="n">plot_bg</span><span class="p">:</span>
                <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">dropna</span><span class="p">(),</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;kinase&quot;</span><span class="p">,</span>
                            <span class="n">palette</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="o">.</span><span class="n">dropna</span><span class="p">()[</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># else remove the unaffected substrates from the plotting df</span>
                <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">bg_col</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(),</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;kinase&quot;</span><span class="p">,</span>
                            <span class="n">palette</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">kseaResults</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">bg_col</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()[</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>

            <span class="c1"># remove top and right spines/plot lines</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;lightgray&quot;</span><span class="p">)</span>
            <span class="c1"># return the figure object only if demanded</span>
            <span class="k">if</span> <span class="n">ret</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
                <span class="k">return</span> <span class="n">fig</span></div>

<div class="viewcode-block" id="KSEA.volcanos"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.volcanos">[docs]</a>    <span class="k">def</span> <span class="nf">volcanos</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">log_fc</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">kinases</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot volcano plots highlighting substrates of a given kinase.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        log_fc : str</span>
<span class="sd">            Column name of column containing the log fold changes.</span>
<span class="sd">            Must be present in the dataframe KSEA was initialised with.</span>
<span class="sd">        p : str</span>
<span class="sd">            Column name of column containing the p values.</span>
<span class="sd">            Must be present in the dataframe KSEA was initialised with.</span>
<span class="sd">        kinases : list of str, optional</span>
<span class="sd">            Limit the analysis to these kinases. The default is [].</span>
<span class="sd">        **kwargs :</span>
<span class="sd">            passed to autoprot.visualisation.volcano.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># generate a df containing only the kinases of interest</span>
        <span class="k">if</span> <span class="n">kinases</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">kinases</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotate_df</span><span class="p">(</span><span class="n">kinases</span><span class="o">=</span><span class="n">kinases</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">kinases</span><span class="p">:</span>
            <span class="c1"># index for highlighting the selected kinase substrates</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
            <span class="n">vis</span><span class="o">.</span><span class="n">volcano</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">log_fc</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">p</span><span class="p">,</span> <span class="n">highlight</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span>
                        <span class="n">custom_hl</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="n">k</span><span class="p">},</span>
                        <span class="n">custom_fg</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="mf">.5</span><span class="p">},</span>
                        <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

<div class="viewcode-block" id="KSEA.return_kinase_substrate"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.return_kinase_substrate">[docs]</a>    <span class="k">def</span> <span class="nf">return_kinase_substrate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinase</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return new dataframe with substrates of one or multiple kinase(s).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinase : str or list of str</span>
<span class="sd">            Kinase(s) to analyse.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If kinase is neither list of str nor str.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        df_filter : pd.Dataframe</span>
<span class="sd">            Dataframe containing detailed information on kinase-substrate pairs</span>
<span class="sd">            including reference literature.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># use the simplified dataset if it is present</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpleDf</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">simpleDf</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># otherwise use the complete dataset including kinase isoforms</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotDf</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># if a list of kinases is provided, iterate through the list and</span>
        <span class="c1"># collect corresponding indices</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;KINASE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="o">==</span> <span class="n">k</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">kinase</span><span class="p">]</span>

            <span class="c1"># merge all row indices and use them to create a sub-df containing</span>
            <span class="c1"># only the kinases of interest</span>
            <span class="n">df_filter</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">idx</span><span class="p">)]</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">df_filter</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;KINASE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="o">==</span> <span class="n">kinase</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Please provide either a string or a list of strings representing kinases of interest.&quot;</span><span class="p">)</span>

        <span class="c1"># data are merged implicitly on common column nnames i.e. on SITE_GRP_ID</span>
        <span class="c1"># only entries present in the filtered annotDfare retained</span>
        <span class="n">df_filter</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_filter</span><span class="p">[[</span><span class="s1">&#39;GENE&#39;</span><span class="p">,</span> <span class="s1">&#39;KINASE&#39;</span><span class="p">,</span> <span class="s1">&#39;KIN_ACC_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;SUBSTRATE&#39;</span><span class="p">,</span> <span class="s1">&#39;SUB_ACC_ID&#39;</span><span class="p">,</span>
                                        <span class="s1">&#39;SUB_GENE&#39;</span><span class="p">,</span> <span class="s1">&#39;SUB_MOD_RSD&#39;</span><span class="p">,</span> <span class="s1">&#39;SITE_GRP_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;SITE_+/-7_AA&#39;</span><span class="p">,</span> <span class="s1">&#39;DOMAIN&#39;</span><span class="p">,</span>
                                        <span class="s1">&#39;IN_VIVO_RXN&#39;</span><span class="p">,</span> <span class="s1">&#39;IN_VITRO_RXN&#39;</span><span class="p">,</span> <span class="s1">&#39;CST_CAT#&#39;</span><span class="p">,</span> <span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s2">&quot;mergeID&quot;</span><span class="p">]],</span>
                             <span class="bp">self</span><span class="o">.</span><span class="n">PSP_regSits</span><span class="p">[[</span><span class="s1">&#39;SITE_GRP_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_FUNCTION&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROT_INTERACT&#39;</span><span class="p">,</span>
                                               <span class="s1">&#39;ON_OTHER_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;PMIDs&#39;</span><span class="p">,</span> <span class="s1">&#39;LT_LIT&#39;</span><span class="p">,</span> <span class="s1">&#39;MS_LIT&#39;</span><span class="p">,</span> <span class="s1">&#39;MS_CST&#39;</span><span class="p">,</span>
                                               <span class="s1">&#39;NOTES&#39;</span><span class="p">]],</span>
                             <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df_filter</span></div>

<div class="viewcode-block" id="KSEA.annotate_df"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.KSEA.annotate_df">[docs]</a>    <span class="k">def</span> <span class="nf">annotate_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinases</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Annotate the provided dataframe with boolean columns for given kinases.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinases : list of str, optional</span>
<span class="sd">            List of kinases. The default is [].</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            annotated dataframe containing a column for each provided kinase</span>
<span class="sd">            with boolean values representing a row/protein being a kinase</span>
<span class="sd">            substrate or not.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">kinases</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">kinases</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kinases</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># remove the two columns from the returned df</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;MOD_RSD&quot;</span><span class="p">,</span> <span class="s2">&quot;ucGene&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">kinases</span><span class="p">:</span>
                <span class="c1"># find substrates for the given kinase in the dataset</span>
                <span class="n">ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">return_kinase_substrate</span><span class="p">(</span><span class="n">kinase</span><span class="p">)[</span><span class="s2">&quot;mergeID&quot;</span><span class="p">]</span>
                <span class="c1"># init the boolean column with zeros</span>
                <span class="n">df</span><span class="p">[</span><span class="n">kinase</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="c1"># check if the unique ID for each protein is present in the</span>
                <span class="c1"># returnKinaseSubstrate df. If so set the column value to 1.</span>
                <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;mergeID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ids</span><span class="p">),</span> <span class="n">kinase</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="c1"># remove also the mergeID column before returning the df</span>
            <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;mergeID&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Please provide kinase(s) for annotation.&quot;</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="miss_analysis"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.miss_analysis">[docs]</a><span class="k">def</span> <span class="nf">miss_analysis</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s1">&#39;ascending&#39;</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                  <span class="n">extra_vis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print missing statistics for a dataframe.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.Dataframe</span>
<span class="sd">        Input dataframe with missing values.</span>
<span class="sd">    cols : list of str</span>
<span class="sd">        Columns to perform missing values analysis on.</span>
<span class="sd">    n : int, optional</span>
<span class="sd">        How many rows of the dataframe to displayed.</span>
<span class="sd">        The default is None (uses all rows).</span>
<span class="sd">    sort : str, optional</span>
<span class="sd">        &quot;ascending&quot; or &quot;descending&quot;.</span>
<span class="sd">        The default is &#39;ascending&#39;.</span>
<span class="sd">    text : bool, optional</span>
<span class="sd">        Whether to output text summaryMap.</span>
<span class="sd">        The default is True.</span>
<span class="sd">    vis : bool, optional</span>
<span class="sd">        whether to return barplot showing missingness.</span>
<span class="sd">        The default is True.</span>
<span class="sd">    extra_vis : bool, optional</span>
<span class="sd">        Whether to return matrix plot showing missingness.</span>
<span class="sd">        The default is False.</span>
<span class="sd">    save_dir : str, optional</span>
<span class="sd">        Path to folder where the results should be saved.</span>
<span class="sd">        The default is None.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If n_entries is incorrectly specified.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    miss_analysis gives a quick overview of the missingness of the provided</span>
<span class="sd">    dataframe. You can provide the complete or prefiltered dataframe as input.</span>
<span class="sd">    Providing n_entries allows you to specify how many of the entries of the dataframe</span>
<span class="sd">    (sorted by missingness) are displayed (i.e. only display the n_entries columns with</span>
<span class="sd">    most (or least) missing values) With the sort argument you can define</span>
<span class="sd">    whether the dataframe is sorted by least to most missing values or vice versa</span>
<span class="sd">    (using &quot;descending&quot; and &quot;ascending&quot;, respectively). The vis and extra_vis</span>
<span class="sd">    arguments can be used to toggle the graphical output.</span>
<span class="sd">    In case of large data (a lot of columns) those might be better turned off.</span>

<span class="sd">    &gt;&gt;&gt; autoprot.analysis.miss_analysis(phos_expanded,</span>
<span class="sd">    ...                                twitchVsctrl+twitchVsmild+mildVsctrl,</span>
<span class="sd">    ...                                sort=&quot;descending&quot;,</span>
<span class="sd">    ...                                extra_vis = True)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.preprocessing as pp</span>
<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import pandas as pd</span>

<span class="sd">        phos = pd.read_csv(&quot;_static/testdata/Phospho (STY)Sites_mod.zip&quot;, sep=&quot;\t&quot;, low_memory=False)</span>
<span class="sd">        phos = pp.cleaning(phos, file = &quot;Phospho (STY)&quot;)</span>
<span class="sd">        phosRatio = phos.filter(regex=&quot;^Ratio .\/.( | normalized )R.___&quot;).columns</span>
<span class="sd">        phos = pp.log(phos, phosRatio, base=2)</span>
<span class="sd">        phos = pp.filter_loc_prob(phos, thresh=.75)</span>
<span class="sd">        phosRatio = phos.filter(regex=&quot;log2_Ratio .\/.( | normalized )R.___&quot;).columns</span>
<span class="sd">        phos = pp.remove_non_quant(phos, phosRatio)</span>

<span class="sd">        phosRatio = phos.filter(regex=&quot;log2_Ratio .\/. normalized R.___&quot;).columns</span>
<span class="sd">        phos_expanded = pp.expand_site_table(phos, phosRatio)</span>

<span class="sd">        twitchVsmild = [&#39;log2_Ratio H/M normalized R1&#39;,&#39;log2_Ratio M/L normalized R2&#39;,&#39;log2_Ratio H/M normalized R3&#39;,</span>
<span class="sd">                        &#39;log2_Ratio H/L normalized R4&#39;,&#39;log2_Ratio H/M normalized R5&#39;,&#39;log2_Ratio M/L normalized R6&#39;]</span>
<span class="sd">        twitchVsctrl = [&quot;log2_Ratio H/L normalized R1&quot;,&quot;log2_Ratio H/M normalized R2&quot;,&quot;log2_Ratio H/L normalized R3&quot;,</span>
<span class="sd">                        &quot;log2_Ratio M/L normalized R4&quot;, &quot;log2_Ratio H/L normalized R5&quot;,&quot;log2_Ratio H/M normalized R6&quot;]</span>
<span class="sd">        mildVsctrl = [&quot;log2_Ratio M/L normalized R1&quot;,&quot;log2_Ratio H/L normalized R2&quot;,&quot;log2_Ratio M/L normalized R3&quot;,</span>
<span class="sd">                      &quot;log2_Ratio H/M normalized R4&quot;,&quot;log2_Ratio M/L normalized R5&quot;,&quot;log2_Ratio H/L normalized R6&quot;]</span>
<span class="sd">        phos = ana.ttest(df=phos_expanded, reps=twitchVsmild, cond=&quot;TvM&quot;)</span>
<span class="sd">        phos = ana.ttest(df=phos_expanded, reps=twitchVsctrl, cond=&quot;TvC&quot;)</span>

<span class="sd">        ana.miss_analysis(phos_expanded,</span>
<span class="sd">                         twitchVsctrl+twitchVsmild+mildVsctrl,</span>
<span class="sd">                         text=False,</span>
<span class="sd">                         sort=&quot;descending&quot;,</span>
<span class="sd">                         extra_vis = True)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># only analyse subset of cols</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span>
    <span class="c1"># sorted list of lists with every sublist containing</span>
    <span class="c1"># [colname,total_n, n_missing, percentage, rank]</span>

    <span class="c1"># calculate summary missing statistics</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># implicitly iterate over dataframe cols</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="p">:</span>
        <span class="c1"># len dataframe</span>
        <span class="n">n_entries</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1"># how many are missing</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="c1"># percentage</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">m</span> <span class="o">/</span> <span class="n">n_entries</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">i</span><span class="p">,</span> <span class="n">n_entries</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">p</span><span class="p">])</span>

    <span class="c1"># Sort data by the percentage of missingness</span>
    <span class="n">data</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">itemgetter</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
    <span class="c1"># inverse dataframe if required</span>
    <span class="k">if</span> <span class="n">sort</span> <span class="o">==</span> <span class="s1">&#39;descending&#39;</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># add a number corresponding to the position in the ranking</span>
    <span class="c1"># to every condition aka column.</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
        <span class="n">col</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

    <span class="c1"># determine number of entries to show</span>
    <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&#39;n_entries&#39; is larger than dataframe!</span><span class="se">\n</span><span class="s2">Displaying complete dataframe.&quot;</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;n_entries&#39; has to be a positive integer!&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">text</span><span class="p">:</span>  <span class="c1"># print summary statistics and saves them to file</span>
        <span class="n">allines</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">allines</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> has </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2"> of </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> entries missing (</span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span> <span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">%).&quot;</span>
            <span class="n">allines</span> <span class="o">+=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
            <span class="c1"># line separator</span>
            <span class="n">allines</span> <span class="o">+=</span> <span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">80</span>

        <span class="k">if</span> <span class="n">save_dir</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_dir</span><span class="si">}</span><span class="s2">/missAnalysis_text.txt&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">allines</span><span class="p">)</span>

        <span class="c1"># write all lines at once</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">allines</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">vis</span><span class="p">:</span>  <span class="c1"># Visualize the % missingness of first n entries of dataframe as a bar plot.</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;tot_values&quot;</span><span class="p">,</span> <span class="s2">&quot;tot_miss&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_miss&quot;</span><span class="p">,</span> <span class="s2">&quot;rank&quot;</span><span class="p">])</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">()</span>
        <span class="c1"># plot colname against total missing values</span>
        <span class="n">splot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;tot_miss&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">n</span><span class="p">],</span>
                            <span class="n">y</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">n</span><span class="p">])</span>

        <span class="c1"># add the percentage of missingness to every bar of the plot</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">splot</span><span class="o">.</span><span class="n">patches</span><span class="p">):</span>
            <span class="n">s</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="mi">2</span><span class="p">))</span><span class="si">}</span><span class="s1">%&#39;</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">get_width</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">get_width</span><span class="p">()</span> <span class="o">*</span> <span class="mf">.01</span>
            <span class="n">y</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">get_y</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">get_height</span><span class="p">()</span> <span class="o">/</span> <span class="mi">2</span>
            <span class="n">splot</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Missing values of dataframe columns.&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">save_dir</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_dir</span><span class="si">}</span><span class="s2">/missAnalysis_vis1.pdf&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">extra_vis</span><span class="p">:</span>  <span class="c1"># Visualize the missingness in the dataset using missingno.</span>
        <span class="c1"># plots are generated with missingno.matrix.</span>
        <span class="c1"># See https://github.com/ResidentMario/missingno</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">msn</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;ascending&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">save_dir</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">+</span> <span class="s2">&quot;/missAnalysis_vis2.pdf&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="get_pub_abstracts"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.get_pub_abstracts">[docs]</a><span class="k">def</span> <span class="nf">get_pub_abstracts</span><span class="p">(</span><span class="n">text</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title_or_abstract</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">author</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">phrase</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                      <span class="n">exclusion</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">custom_search_term</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">make_word_cloud</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                      <span class="n">sort</span><span class="o">=</span><span class="s1">&#39;pub+date&#39;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="s2">&quot;print&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get Pubmed abstracts.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Provide the search terms and pubmed will be searched for paper matching those</span>
<span class="sd">    The exclusion list can be used to exclude certain words to be contained in a paper</span>
<span class="sd">    If you want to perform more complicated searches use the possibility of using</span>
<span class="sd">    a custom term.</span>

<span class="sd">    If you search for a not excact term you can use *</span>
<span class="sd">    For example opto* will also return optogenetic.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    text : list of str, optional</span>
<span class="sd">        List of words which must occur in text.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    title_or_abstract : list of str, optional</span>
<span class="sd">        List of words which must occur in Titel or Abstract.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    author : list of str, optional</span>
<span class="sd">        List of authors which must occor in Author list.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    phrase : list of str, optional</span>
<span class="sd">        List of phrases which should occur in text.</span>
<span class="sd">        Seperate word in phrases with hyphen (e.g. akt-signaling).</span>
<span class="sd">        The default is None.</span>
<span class="sd">    exclusion : list of str, optional</span>
<span class="sd">        List of words which must not occur in text.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    custom_search_term : str, optional</span>
<span class="sd">        Enter a custom search term (make use of advanced pubmed search functionality).</span>
<span class="sd">        If this is supplied all other search terms will be ignored.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    make_word_cloud : bool, optional</span>
<span class="sd">        Whether to draw a wordcloud of the words in retrieved abstrace.</span>
<span class="sd">        The default is False.</span>
<span class="sd">    sort : str, optional</span>
<span class="sd">        How to sort the results.</span>
<span class="sd">        The default is &#39;pub+date&#39;.</span>
<span class="sd">    retmax : int, optional</span>
<span class="sd">        Maximum number of found articles to return.</span>
<span class="sd">        The default is 20.</span>
<span class="sd">    output : str, optional</span>
<span class="sd">        How to handle the output.</span>
<span class="sd">        Possible values are &#39;SOMEPATH.txt&#39;,</span>
<span class="sd">        &#39;SOMEPATH.html&#39; or &#39;SOMEPATH&#39;.</span>
<span class="sd">        If no extension is given, the output will be html.</span>
<span class="sd">        The default is &quot;print&quot;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    To generate a wordcloud and print the results of the found articles to the</span>
<span class="sd">    prompt use the following command</span>

<span class="sd">    &gt;&gt;&gt; autoprot.analysis.get_pub_abstracts(titel_or_abstract=[&quot;p38&quot;, &quot;JNK&quot;, &quot;ERK&quot;],</span>
<span class="sd">                                          make_word_cloud=True)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        ana.get_pub_abstracts(title_or_abstract=[&quot;p38&quot;, &quot;JNK&quot;, &quot;ERK&quot;],</span>
<span class="sd">                              make_word_cloud=True)</span>

<span class="sd">    Even more comfortably, you can also save the results incl. the wordcloud</span>
<span class="sd">    as html file</span>

<span class="sd">    &gt;&gt;&gt;  autoprot.analysis.get_pub_abstracts(titel_or_abstract=[&quot;p38&quot;, &quot;JNK&quot;, &quot;ERK&quot;],</span>
<span class="sd">                                          make_word_cloud=True,</span>
<span class="sd">                                          output=&#39;./MyPubmedSearch.html&#39;)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">exclusion</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">exclusion</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">phrase</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">phrase</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">author</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">author</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">title_or_abstract</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">title_or_abstract</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">text</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">text</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>

    <span class="c1"># STEP 1: Generate the search term</span>
    <span class="k">if</span> <span class="n">custom_search_term</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># generate a pubmed search term from user input</span>
        <span class="c1"># Generate long list of concatenated search terms</span>
        <span class="n">term</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="s2">&quot;[Title/Abstract]&quot;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">title_or_abstract</span><span class="p">]</span> <span class="o">+</span> \
               <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="s2">&quot;[Text Word]&quot;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">text</span><span class="p">]</span> <span class="o">+</span> \
               <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="s2">&quot;[Author]&quot;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">author</span><span class="p">]</span> <span class="o">+</span> \
               <span class="p">[</span><span class="n">i</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">phrase</span><span class="p">]</span>
        <span class="n">term</span> <span class="o">=</span> <span class="s2">&quot; AND &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">term</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">])</span>

        <span class="c1"># add exclusions joined by NOT</span>
        <span class="k">if</span> <span class="n">exclusion</span> <span class="o">!=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]:</span>
            <span class="n">exclusion</span> <span class="o">=</span> <span class="s2">&quot; NOT &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exclusion</span><span class="p">)</span>
            <span class="n">term</span> <span class="o">=</span> <span class="s2">&quot; NOT &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">term</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">exclusion</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># if the user provides a custom search term, ignore all remaining input</span>
        <span class="n">term</span> <span class="o">=</span> <span class="n">custom_search_term</span>

    <span class="c1"># STEP 2: Perform the PubMed search and return the PubMed IDs of the found articles</span>
    <span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s1">&#39;your.email@example.com&#39;</span>  <span class="c1"># seemingly no true mail address is required.</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s1">&#39;pubmed&#39;</span><span class="p">,</span>
                            <span class="n">sort</span><span class="o">=</span><span class="n">sort</span><span class="p">,</span>
                            <span class="n">retmax</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">retmax</span><span class="p">),</span>
                            <span class="n">retmode</span><span class="o">=</span><span class="s1">&#39;xml&#39;</span><span class="p">,</span>
                            <span class="n">term</span><span class="o">=</span><span class="n">term</span><span class="p">)</span>
    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mf">0.5</span><span class="p">)</span>  <span class="c1"># if function is executed in a loop: a small delay to ensure pubmed is responding</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>

    <span class="c1"># if the search was not successful, leave the function</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No results for &quot;</span> <span class="o">+</span> <span class="n">term</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># STEP 3: get more detailed results including abstracts, authors etc</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s1">&#39;pubmed&#39;</span><span class="p">,</span>
                           <span class="n">retmode</span><span class="o">=</span><span class="s1">&#39;xml&#39;</span><span class="p">,</span>
                           <span class="nb">id</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">results</span><span class="p">)))</span>
    <span class="n">papers</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)[</span><span class="s2">&quot;PubmedArticle&quot;</span><span class="p">]</span>

    <span class="n">titles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">abstracts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">authors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">dois</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">links</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">pids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">dates</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">journals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">final_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="c1"># iterate through the papers</span>
    <span class="k">for</span> <span class="n">paper</span> <span class="ow">in</span> <span class="n">papers</span><span class="p">:</span>
        <span class="c1"># get titles</span>
        <span class="n">titles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">paper</span><span class="p">[</span><span class="s2">&quot;MedlineCitation&quot;</span><span class="p">][</span><span class="s2">&quot;Article&quot;</span><span class="p">][</span><span class="s2">&quot;ArticleTitle&quot;</span><span class="p">])</span>
        <span class="c1"># get abstract</span>
        <span class="k">if</span> <span class="s2">&quot;Abstract&quot;</span> <span class="ow">in</span> <span class="n">paper</span><span class="p">[</span><span class="s2">&quot;MedlineCitation&quot;</span><span class="p">][</span><span class="s2">&quot;Article&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">abstracts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">paper</span><span class="p">[</span><span class="s2">&quot;MedlineCitation&quot;</span><span class="p">][</span><span class="s2">&quot;Article&quot;</span><span class="p">][</span><span class="s2">&quot;Abstract&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">abstracts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;No abstract&quot;</span><span class="p">)</span>
        <span class="c1"># get authors</span>
        <span class="n">al</span> <span class="o">=</span> <span class="n">paper</span><span class="p">[</span><span class="s1">&#39;MedlineCitation&#39;</span><span class="p">][</span><span class="s1">&#39;Article&#39;</span><span class="p">][</span><span class="s2">&quot;AuthorList&quot;</span><span class="p">]</span>
        <span class="n">temp_authors</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">al</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;ForeName&quot;</span> <span class="ow">in</span> <span class="n">a</span> <span class="ow">and</span> <span class="s2">&quot;LastName&quot;</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
                <span class="n">temp_authors</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">a</span><span class="p">[</span><span class="s1">&#39;ForeName&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">a</span><span class="p">[</span><span class="s1">&#39;LastName&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
            <span class="k">elif</span> <span class="s2">&quot;LastName&quot;</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
                <span class="n">temp_authors</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">a</span><span class="p">[</span><span class="s1">&#39;LastName&#39;</span><span class="p">]])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">temp_authors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
        <span class="n">authors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">temp_authors</span><span class="p">)))</span>
        <span class="c1"># get dois and make link</span>
        <span class="n">doi</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">paper</span><span class="p">[</span><span class="s1">&#39;PubmedData&#39;</span><span class="p">][</span><span class="s1">&#39;ArticleIdList&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;IdType&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;doi&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">doi</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">doi</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">doi</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">doi</span> <span class="o">=</span> <span class="s2">&quot;NaN&quot;</span>
        <span class="n">dois</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">doi</span><span class="p">)</span>
        <span class="n">links</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot; https://doi.org/</span><span class="si">{</span><span class="n">doi</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># get pids</span>
        <span class="n">pids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">paper</span><span class="p">[</span><span class="s1">&#39;PubmedData&#39;</span><span class="p">][</span><span class="s1">&#39;ArticleIdList&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;IdType&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;pubmed&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span>
        <span class="c1"># get dates</span>

        <span class="n">raw_date</span> <span class="o">=</span> <span class="n">paper</span><span class="p">[</span><span class="s1">&#39;MedlineCitation&#39;</span><span class="p">][</span><span class="s1">&#39;Article&#39;</span><span class="p">][</span><span class="s1">&#39;Journal&#39;</span><span class="p">][</span><span class="s1">&#39;JournalIssue&#39;</span><span class="p">][</span><span class="s1">&#39;PubDate&#39;</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">date</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">raw_date</span><span class="p">[</span><span class="s1">&#39;Year&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">raw_date</span><span class="p">[</span><span class="s1">&#39;Month&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">raw_date</span><span class="p">[</span><span class="s1">&#39;Day&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">date</span> <span class="o">=</span> <span class="n">raw_date</span>
        <span class="n">dates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">date</span><span class="p">)</span>
        <span class="c1"># get journal</span>
        <span class="n">journals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">paper</span><span class="p">[</span><span class="s1">&#39;MedlineCitation&#39;</span><span class="p">][</span><span class="s1">&#39;Article&#39;</span><span class="p">][</span><span class="s1">&#39;Journal&#39;</span><span class="p">][</span><span class="s1">&#39;Title&#39;</span><span class="p">])</span>

    <span class="c1"># sum up the results in a dict containing the search term as key</span>
    <span class="n">final_dict</span><span class="p">[</span><span class="n">term</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">titles</span><span class="p">,</span> <span class="n">authors</span><span class="p">,</span> <span class="n">abstracts</span><span class="p">,</span> <span class="n">dois</span><span class="p">,</span> <span class="n">links</span><span class="p">,</span> <span class="n">pids</span><span class="p">,</span> <span class="n">dates</span><span class="p">,</span> <span class="n">journals</span><span class="p">)</span>

    <span class="c1"># STEP 4: Make a word cloud picture from Pubmed search abstracts.</span>
    <span class="k">if</span> <span class="n">make_word_cloud</span><span class="p">:</span>  <span class="c1"># plot a picture of words in the found abstracts</span>
        <span class="n">abstracts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># for every search term (there usually is only one)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">final_dict</span><span class="p">:</span>
            <span class="c1"># the abstract is the third element in the list</span>
            <span class="n">abstracts</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="n">abstract</span><span class="p">[</span><span class="s2">&quot;AbstractText&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">abstract</span> <span class="ow">in</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">abstract</span><span class="p">,</span> <span class="nb">str</span><span class="p">))</span>
        <span class="c1"># create a long string of abstract words</span>
        <span class="n">abstracts</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">abstracts</span><span class="p">)</span>

        <span class="c1"># when exlusion list is added in wordcloud add those</span>
        <span class="c1"># TODO properly implement the inclusion of exclusions</span>
        <span class="n">el</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;sup&quot;</span><span class="p">]</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="n">vis</span><span class="o">.</span><span class="n">wordcloud</span><span class="p">(</span><span class="n">text</span><span class="o">=</span><span class="n">abstracts</span><span class="p">,</span> <span class="n">exlusion_words</span><span class="o">=</span><span class="n">el</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output</span> <span class="o">!=</span> <span class="s2">&quot;print&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">output</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.txt&quot;</span><span class="p">:</span>
                <span class="n">output</span> <span class="o">=</span> <span class="n">output</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">to_file</span><span class="p">(</span><span class="n">output</span> <span class="o">+</span> <span class="s2">&quot;/WordCloud.png&quot;</span><span class="p">)</span>

    <span class="c1"># STEP 5: generate a txt or html output or print to the prompt</span>
    <span class="k">if</span> <span class="n">output</span> <span class="o">==</span> <span class="s2">&quot;print&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">final_dict</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">author</span><span class="p">,</span> <span class="n">abstract</span><span class="p">,</span> <span class="n">doi</span><span class="p">,</span> <span class="n">link</span><span class="p">,</span> <span class="n">pid</span><span class="p">,</span> <span class="n">date</span><span class="p">,</span> <span class="n">journal</span> <span class="ow">in</span> \
                    <span class="nb">zip</span><span class="p">(</span><span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">4</span><span class="p">],</span>
                        <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">5</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">6</span><span class="p">],</span>
                        <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">7</span><span class="p">]):</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">author</span><span class="p">))</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">journal</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">date</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;doi: </span><span class="si">{</span><span class="n">doi</span><span class="si">}</span><span class="s2">, PubMedID: </span><span class="si">{</span><span class="n">pid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">link</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">abstract</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">abstract</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">abstract</span><span class="p">[</span><span class="s2">&quot;AbstractText&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">output</span><span class="p">[</span><span class="o">-</span><span class="mi">4</span><span class="p">:]</span> <span class="o">==</span> <span class="s2">&quot;.txt&quot;</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">final_dict</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">author</span><span class="p">,</span> <span class="n">abstract</span><span class="p">,</span> <span class="n">doi</span><span class="p">,</span> <span class="n">link</span><span class="p">,</span> <span class="n">pid</span><span class="p">,</span> <span class="n">date</span><span class="p">,</span> <span class="n">journal</span> <span class="ow">in</span> \
                        <span class="nb">zip</span><span class="p">(</span><span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span>
                            <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">4</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">5</span><span class="p">],</span>
                            <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">6</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">7</span><span class="p">]):</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">author</span><span class="p">))</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">journal</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">date</span><span class="p">))</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;doi: </span><span class="si">{</span><span class="n">doi</span><span class="si">}</span><span class="s2">, PubMedID: </span><span class="si">{</span><span class="n">pid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">link</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

                    <span class="c1"># write the abstract in a structured manner with defined line breaks</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">abstract</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">abstract</span> <span class="o">=</span> <span class="n">abstract</span><span class="p">[</span><span class="s2">&quot;AbstractText&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">abstract</span> <span class="o">=</span> <span class="n">abstract</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">word</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">abstract</span><span class="p">):</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">word</span><span class="p">)</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">idx</span> <span class="o">%</span> <span class="mi">20</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">idx</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># if the output is a html file or a folder, write html</span>
    <span class="k">elif</span> <span class="n">output</span><span class="p">[</span><span class="o">-</span><span class="mi">5</span><span class="p">:]</span> <span class="o">==</span> <span class="s2">&quot;.html&quot;</span> <span class="ow">or</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output</span><span class="p">):</span>
            <span class="n">output</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s2">&quot;PubCrawlerResults.html&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;!DOC html&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;.center {&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;display: block;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;margin-left: auto;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;margin-right: auto;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;width: 80%;}&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;body style=&quot;background-color:#FFE5B4&quot;&gt;&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">make_word_cloud</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;img src=&quot;WordCloud.png&quot; alt=&quot;WordCloud&quot; class=&quot;center&quot;&gt;&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">final_dict</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">author</span><span class="p">,</span> <span class="n">abstract</span><span class="p">,</span> <span class="n">doi</span><span class="p">,</span> <span class="n">link</span><span class="p">,</span> <span class="n">pid</span><span class="p">,</span> <span class="n">date</span><span class="p">,</span> <span class="n">journal</span> <span class="ow">in</span> \
                        <span class="nb">zip</span><span class="p">(</span><span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span>
                            <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">4</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">5</span><span class="p">],</span>
                            <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">6</span><span class="p">],</span> <span class="n">final_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">7</span><span class="p">]):</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">ta</span> <span class="o">=</span> <span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">author</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;p style=&quot;color:gray; font-size:16px&quot;&gt;</span><span class="si">{</span><span class="n">ta</span><span class="si">}</span><span class="s1">&lt;/p&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">200</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;i&gt;</span><span class="si">{</span><span class="n">journal</span><span class="si">}</span><span class="s2">&lt;/i&gt;&quot;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">date</span><span class="p">))</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;i&gt;doi:&lt;/i&gt; </span><span class="si">{</span><span class="n">doi</span><span class="si">}</span><span class="s2">, &lt;i&gt;PubMedID:&lt;/i&gt; </span><span class="si">{</span><span class="n">pid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;a href=</span><span class="si">{</span><span class="n">link</span><span class="si">}</span><span class="s1">&gt;&lt;i&gt;Link to journal&lt;/i&gt;&lt;/a&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">200</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>

                    <span class="c1"># write the abstract in a structured manner with defined line breaks</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">abstract</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">abstract</span> <span class="o">=</span> <span class="n">abstract</span><span class="p">[</span><span class="s2">&quot;AbstractText&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">abstract</span> <span class="o">=</span> <span class="n">abstract</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;p style=&quot;color:#202020&quot;&gt;&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">word</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">abstract</span><span class="p">):</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">word</span><span class="p">)</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">idx</span> <span class="o">%</span> <span class="mi">20</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">idx</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;/p&gt;&#39;</span><span class="p">)</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;p style=&quot;color:#202020&quot;&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;/p&gt;&#39;</span><span class="p">)</span>

                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">200</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;*&#39;</span> <span class="o">*</span> <span class="mi">150</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">200</span><span class="p">)</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&lt;br&gt;&#39;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="loess"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.loess">[docs]</a><span class="k">def</span> <span class="nf">loess</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">xvals</span><span class="p">,</span> <span class="n">yvals</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">poly_degree</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate a LOcally-Weighted Scatterplot Smoothing Fit.</span>

<span class="sd">    See: https://medium.com/@langen.mu/creating-powerfull-lowess-graphs-in-python-e0ea7a30b17a</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data : pd.Dataframe</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    xvals : str</span>
<span class="sd">        Colname of x values.</span>
<span class="sd">    yvals : str</span>
<span class="sd">        Colname of y values.</span>
<span class="sd">    alpha : float</span>
<span class="sd">        Sensitivity of the estimation.</span>
<span class="sd">        Controls how much of the total number of values is used during weighing.</span>
<span class="sd">        0 &lt;= alpha &lt;= 1.</span>
<span class="sd">    poly_degree : int, optional</span>
<span class="sd">        Degree of the fitted polynomial. The default is 2.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    Loess normalisation (also referred to as Savitzky-Golay filter) locally approximates</span>
<span class="sd">    the data around every point using low-order functions and giving less weight to distant</span>
<span class="sd">    data points.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; np.random.seed(10)</span>
<span class="sd">    &gt;&gt;&gt; x_values = np.random.randint(-50,110,size=250)</span>
<span class="sd">    &gt;&gt;&gt; y_values = np.square(x_values)/1.5 + np.random.randint(-1000,1000, size=len(x_values))</span>
<span class="sd">    &gt;&gt;&gt; df = pd.DataFrame({&quot;Xvalue&quot; : x_values,</span>
<span class="sd">                           &quot;Yvalue&quot; : y_values</span>
<span class="sd">                           })</span>

<span class="sd">    &gt;&gt;&gt; evalDF = autoprot.analysis.loess(df, &quot;Xvalue&quot;, &quot;Yvalue&quot;, alpha=0.7, poly_degree=2)</span>
<span class="sd">    &gt;&gt;&gt; fig, ax = plt.subplots(1,1)</span>
<span class="sd">    &gt;&gt;&gt; sns.scatterplot(df[&quot;Xvalue&quot;], df[&quot;Yvalue&quot;], ax=ax)</span>
<span class="sd">    &gt;&gt;&gt; ax.plot(eval_df[&#39;v&#39;], eval_df[&#39;g&#39;], color=&#39;red&#39;, linewidth= 3, label=&quot;Test&quot;)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>
<span class="sd">        import seaborn as sns</span>

<span class="sd">        x_values = np.random.randint(-50,110,size=(250))</span>
<span class="sd">        y_values = np.square(x_values)/1.5 + np.random.randint(-1000,1000, size=len(x_values))</span>
<span class="sd">        df = pd.DataFrame({&quot;Xvalue&quot; : x_values,</span>
<span class="sd">                           &quot;Yvalue&quot; : y_values</span>
<span class="sd">                           })</span>
<span class="sd">        evalDF = ana.loess(df, &quot;Xvalue&quot;, &quot;Yvalue&quot;, alpha=0.7, poly_degree=2)</span>
<span class="sd">        fig, ax = plt.subplots(1,1)</span>
<span class="sd">        sns.scatterplot(df[&quot;Xvalue&quot;], df[&quot;Yvalue&quot;], ax=ax)</span>
<span class="sd">        ax.plot(evalDF[&#39;v&#39;], evalDF[&#39;g&#39;], color=&#39;red&#39;, linewidth= 3, label=&quot;Test&quot;)</span>
<span class="sd">        plt.show()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># generate x,y value pairs and sort them according to x</span>
    <span class="n">all_data</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">xvals</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">data</span><span class="p">[</span><span class="n">yvals</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1"># separate the values again into x and y cols</span>
    <span class="n">xvals</span><span class="p">,</span> <span class="n">yvals</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">all_data</span><span class="p">)</span>
    <span class="c1"># generate empty df for final fit</span>
    <span class="n">eval_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">])</span>

    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">xvals</span><span class="p">)</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="c1"># how many data points to include in the weighing</span>
    <span class="c1"># alpha determines the relative proportion of values considered during weighing</span>
    <span class="n">q</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="n">alpha</span><span class="p">)</span> <span class="k">if</span> <span class="n">alpha</span> <span class="o">&lt;=</span> <span class="mf">1.0</span> <span class="k">else</span> <span class="n">n</span><span class="p">)</span>
    <span class="c1"># the average point to point distance in x direction</span>
    <span class="n">avg_interval</span> <span class="o">=</span> <span class="p">((</span><span class="nb">max</span><span class="p">(</span><span class="n">xvals</span><span class="p">)</span> <span class="o">-</span> <span class="nb">min</span><span class="p">(</span><span class="n">xvals</span><span class="p">))</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">xvals</span><span class="p">))</span>
    <span class="c1"># calculate upper on lower boundaries</span>
    <span class="n">v_lb</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">xvals</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="mf">.5</span> <span class="o">*</span> <span class="n">avg_interval</span><span class="p">)</span>
    <span class="n">v_ub</span> <span class="o">=</span> <span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">xvals</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="mf">.5</span> <span class="o">*</span> <span class="n">avg_interval</span><span class="p">))</span>
    <span class="c1"># coordinates for the fitting points</span>
    <span class="n">v</span> <span class="o">=</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">v_lb</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="n">v_ub</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="n">m</span><span class="p">),</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># create an array of ones of the same length as xvals</span>
    <span class="n">xcols</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">xvals</span><span class="p">)]</span>

    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="p">(</span><span class="n">poly_degree</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)):</span>
        <span class="n">xcols</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">i</span> <span class="o">**</span> <span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xvals</span><span class="p">])</span>
    <span class="n">x_mtx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">xcols</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">v</span><span class="p">:</span>
        <span class="n">iterval</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">iterdists</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([(</span><span class="n">j</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">j</span> <span class="o">-</span> <span class="n">iterval</span><span class="p">))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">xvals</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">raw_dists</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">iterdists</span><span class="p">)</span>
        <span class="n">scale_fact</span> <span class="o">=</span> <span class="n">raw_dists</span><span class="p">[</span><span class="n">q</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">scaled_dists</span> <span class="o">=</span> <span class="p">[(</span><span class="n">j</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="n">j</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">scale_fact</span><span class="p">))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">iterdists</span><span class="p">]</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="p">[(</span><span class="n">j</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">((</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">j</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">3</span><span class="p">))</span> <span class="o">**</span> <span class="mi">3</span> <span class="k">if</span> <span class="n">j</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">scaled_dists</span><span class="p">]</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">raw_dists</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="n">iterdists</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">scaled_dists</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="n">scaled_dists</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
        <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">x_mtx</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">w</span> <span class="o">@</span> <span class="n">x_mtx</span><span class="p">)</span> <span class="o">@</span> <span class="p">(</span><span class="n">x_mtx</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">w</span> <span class="o">@</span> <span class="n">yvals</span><span class="p">)</span>
        <span class="c1"># loc_eval</span>
        <span class="n">local_est</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">iterval</span> <span class="o">**</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">b</span><span class="p">))</span>
        <span class="n">iter_df2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;v&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">iterval</span><span class="p">],</span>
            <span class="s1">&#39;g&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">local_est</span><span class="p">]</span>
        <span class="p">})</span>
        <span class="n">eval_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">eval_df</span><span class="p">,</span> <span class="n">iter_df2</span><span class="p">])</span>
    <span class="n">eval_df</span> <span class="o">=</span> <span class="n">eval_df</span><span class="p">[[</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;g&#39;</span><span class="p">]]</span>
    <span class="k">return</span> <span class="n">eval_df</span></div>


<div class="viewcode-block" id="edm"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.edm">[docs]</a><span class="k">def</span> <span class="nf">edm</span><span class="p">(</span><span class="n">matrix_a</span><span class="p">,</span> <span class="n">matrix_b</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate an euclidean distance matrix between two matrices.</span>

<span class="sd">    See:  https://medium.com/swlh/euclidean-distance-matrix-4c3e1378d87f</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix_a : np.ndarray</span>
<span class="sd">        Matrix 1.</span>
<span class="sd">    matrix_b : np.ndarray</span>
<span class="sd">        Matrix 2.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        Distance matrix.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">p1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">matrix_a</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
    <span class="n">p2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">matrix_b</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">p3</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">matrix_a</span><span class="p">,</span> <span class="n">matrix_b</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">p1</span> <span class="o">+</span> <span class="n">p2</span> <span class="o">+</span> <span class="n">p3</span><span class="p">)</span></div>


<div class="viewcode-block" id="limma"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.limma">[docs]</a><span class="k">def</span> <span class="nf">limma</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">reps</span><span class="p">,</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">custom_design</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">calc_contrasts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">print_r</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="c1"># sourcery skip: extract-method, inline-immediately-returned-variable</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform moderated ttest as implemented from R LIMMA.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    reps : list of lists of str</span>
<span class="sd">        Column names of the replicates.</span>
<span class="sd">        Common replicates are grouped together in a single list.</span>
<span class="sd">    cond : str, optional</span>
<span class="sd">        Term to append to the newly generated colnames.</span>
<span class="sd">        The default is &quot;&quot;.</span>
<span class="sd">    custom_design : str, optional</span>
<span class="sd">        Path to custom design file.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    calc_contrasts : str, optional</span>
<span class="sd">        The contrasts to be calculated by limma. Must refer to the column names</span>
<span class="sd">        in the data file. Differences are indicated e.g. by &quot;CondA-CondB&quot;.</span>
<span class="sd">        The autoprot implementation only calculates contrasts based on the</span>
<span class="sd">        column names specified in the design matrix.</span>
<span class="sd">        The default is None.</span>
<span class="sd">    print_r : bool, optional</span>
<span class="sd">        Whether to print the R output.</span>
<span class="sd">        The default is False.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        The input dataframe with additional columns.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    matrix_a custom design representing the design matrix of the microarray experiment,</span>
<span class="sd">    with rows corresponding to arrays and columns to coefficients to be estimated</span>
<span class="sd">    can be provided using customDesign.</span>
<span class="sd">    If customDesign is the unit vector meaning that the arrays are treated as replicates.</span>
<span class="sd">    See: https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/lmFit</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; data = pd.DataFrame({&quot;a1&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                      &quot;a2&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                      &quot;a3&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">    ...                      &quot;b1&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">    ...                      &quot;b2&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">    ...                      &quot;b3&quot;:np.random.normal(loc=0.5, size=4000),})</span>
<span class="sd">    &gt;&gt;&gt; testRes = ana.limma(df=data, reps=[[&quot;a1&quot;,&quot;a2&quot;, &quot;a3&quot;],[&quot;b1&quot;,&quot;b2&quot;, &quot;b3&quot;]], cond=&quot;_test&quot;)</span>
<span class="sd">    &gt;&gt;&gt; testRes[&quot;P.Value_test&quot;].hist()</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import autoprot.analysis as ana</span>

<span class="sd">        df = pd.DataFrame({&quot;a1&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                           &quot;a2&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                           &quot;a3&quot;:np.random.normal(loc=0, size=4000),</span>
<span class="sd">                           &quot;b1&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">                           &quot;b2&quot;:np.random.normal(loc=0.5, size=4000),</span>
<span class="sd">                           &quot;b3&quot;:np.random.normal(loc=0.5, size=4000),})</span>
<span class="sd">        testRes = ana.limma(df, reps=[[&quot;a1&quot;,&quot;a2&quot;, &quot;a3&quot;],[&quot;b1&quot;,&quot;b2&quot;, &quot;b3&quot;]], cond=&quot;_test&quot;)</span>
<span class="sd">        testRes[&quot;P.Value_test&quot;].hist()</span>
<span class="sd">        plt.show()</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># TODO: better handle coefficient extraction in R</span>
    <span class="n">d</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
    <span class="n">data_loc</span> <span class="o">=</span> <span class="n">d</span> <span class="o">+</span> <span class="s2">&quot;/input.csv&quot;</span>
    <span class="n">output_loc</span> <span class="o">=</span> <span class="n">d</span> <span class="o">+</span> <span class="s2">&quot;/output.csv&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;UID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;UID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># flatten in case of two sample</span>
    <span class="n">pp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">df</span><span class="p">[[</span><span class="s2">&quot;UID&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">reps</span><span class="p">))],</span> <span class="n">data_loc</span><span class="p">)</span>

    <span class="c1"># Normally no custom_design is provided</span>
    <span class="k">if</span> <span class="n">custom_design</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">design_loc</span> <span class="o">=</span> <span class="n">d</span> <span class="o">+</span> <span class="s2">&quot;/design.csv&quot;</span>
        <span class="c1"># if two lists are provided with reps, this likely is a twoSample test</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LIMMA: Assuming a two sample test with:&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample 1: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]])))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample 2: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]])))</span>
            <span class="n">test</span> <span class="o">=</span> <span class="s2">&quot;twoSample&quot;</span>
            <span class="n">design</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Intercept&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">])),</span>
                                   <span class="s2">&quot;coef&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">])})</span>
            <span class="c1"># =============================================================================</span>
            <span class="c1">#             creates a design matrix such as</span>
            <span class="c1">#                 Intercept  coef</span>
            <span class="c1">#              0          1     0</span>
            <span class="c1">#              1          1     0</span>
            <span class="c1">#              2          1     0</span>
            <span class="c1">#              3          1     1</span>
            <span class="c1">#              4          1     1</span>
            <span class="c1">#              5          1     1</span>
            <span class="c1"># =============================================================================</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using design matrix:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">design</span><span class="o">.</span><span class="n">to_markdown</span><span class="p">())</span>

            <span class="c1"># save the design for R to read</span>
            <span class="n">pp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">design</span><span class="p">,</span> <span class="n">design_loc</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">test</span> <span class="o">=</span> <span class="s2">&quot;oneSample&quot;</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LIMMA: Assuming a one sample test&quot;</span><span class="p">)</span>
            <span class="c1"># The R function will generate a design matrix corresponding to</span>
            <span class="c1"># ones</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LIMMA: Assuming a custom design test with:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Design specified at </span><span class="si">{</span><span class="n">custom_design</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Columns: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">reps</span><span class="p">)))))</span>

        <span class="n">design</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">custom_design</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using design matrix:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">design</span><span class="o">.</span><span class="n">to_markdown</span><span class="p">())</span>

        <span class="n">test</span> <span class="o">=</span> <span class="s2">&quot;custom&quot;</span>
        <span class="n">design_loc</span> <span class="o">=</span> <span class="n">custom_design</span>

    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">R</span><span class="p">,</span> <span class="s1">&#39;--vanilla&#39;</span><span class="p">,</span> <span class="n">RFUNCTIONS</span><span class="p">,</span> <span class="s2">&quot;limma&quot;</span><span class="p">,</span> <span class="n">data_loc</span><span class="p">,</span> <span class="n">output_loc</span><span class="p">,</span> <span class="n">test</span><span class="p">,</span> <span class="n">design_loc</span><span class="p">,</span> <span class="n">calc_contrasts</span> <span class="ow">or</span> <span class="s2">&quot;&quot;</span><span class="p">]</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span>
            <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">stderr</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">print_r</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

    <span class="n">res</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_loc</span><span class="p">)</span>
    <span class="n">res</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="n">cond</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;UID&quot;</span> <span class="k">else</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">res</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="c1"># this keeps the index of the original df in the returned df</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;UID&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>

    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">data_loc</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">output_loc</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">custom_design</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">design_loc</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="rank_prod"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.rank_prod">[docs]</a><span class="k">def</span> <span class="nf">rank_prod</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">reps</span><span class="p">,</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">print_r</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">correct_fc</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform RankProd test as in R RankProd package.</span>

<span class="sd">    At the moment one sample test only.</span>
<span class="sd">    Test for up and downregulated genes separatly therefore returns two p values.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe.</span>
<span class="sd">    reps : list of lists of str</span>
<span class="sd">        Column names of the replicates.</span>
<span class="sd">        Common replicates are grouped together in a single list.</span>
<span class="sd">    cond : str, optional</span>
<span class="sd">        Term to append to the newly generated colnames.</span>
<span class="sd">        The default is &quot;&quot;.</span>
<span class="sd">    print_r : bool, optional</span>
<span class="sd">        Whether to print the R output.</span>
<span class="sd">        The default is False.</span>
<span class="sd">    correct_fc : bool, optional</span>
<span class="sd">        The rankProd package does not calculate fold-changes for rows with</span>
<span class="sd">        missing values (p Values are calculated). If correct_fc is False</span>
<span class="sd">        the original fold changes from rankProd are return, else fold</span>
<span class="sd">        changes are calculated for all values after ignoring NaNs.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        Input dataframe with additional columns from RankProd.</span>


<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The adjusted p-values returned from the R backend are the percentage of</span>
<span class="sd">    false postives calculated by the rankProd package. This is akin to corrected</span>
<span class="sd">    p values, but care should be taken to name these values accordingly.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">d</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
    <span class="n">data_loc</span> <span class="o">=</span> <span class="n">d</span> <span class="o">+</span> <span class="s2">&quot;/input.csv&quot;</span>
    <span class="n">output_loc</span> <span class="o">=</span> <span class="n">d</span> <span class="o">+</span> <span class="s2">&quot;/output.csv&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;UID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;UID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">class_labels</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;rankProd: Assuming a two sample test with:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample 1: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]])))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample 2: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]])))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Class labels: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">class_labels</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;rankProd: Assuming a one sample test&quot;</span><span class="p">)</span>
        <span class="n">class_labels</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span>

    <span class="c1"># flatten in case of two sample</span>
    <span class="n">pp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">df</span><span class="p">[[</span><span class="s2">&quot;UID&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">reps</span><span class="p">))],</span> <span class="n">data_loc</span><span class="p">)</span>

    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">R</span><span class="p">,</span> <span class="s1">&#39;--vanilla&#39;</span><span class="p">,</span>
               <span class="n">RFUNCTIONS</span><span class="p">,</span>  <span class="c1"># script location</span>
               <span class="s2">&quot;rankProd&quot;</span><span class="p">,</span>  <span class="c1"># functionName</span>
               <span class="n">data_loc</span><span class="p">,</span>  <span class="c1"># data location</span>
               <span class="n">output_loc</span><span class="p">,</span>  <span class="c1"># output file,</span>
               <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">class_labels</span><span class="p">]),</span>
               <span class="p">]</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span>
            <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">stderr</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">print_r</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

    <span class="n">res</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_loc</span><span class="p">)</span>
    <span class="n">res</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="n">cond</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="s2">&quot;UID&quot;</span> <span class="k">else</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">res</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;UID&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">correct_fc</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;log_fc&#39;</span> <span class="o">+</span> <span class="n">cond</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">skipna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">skipna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;log_fc&#39;</span> <span class="o">+</span> <span class="n">cond</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reps</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">data_loc</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">output_loc</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="annotate_phosphosite"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.annotate_phosphosite">[docs]</a><span class="k">def</span> <span class="nf">annotate_phosphosite</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">ps</span><span class="p">,</span> <span class="n">cols_to_keep</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Annotate phosphosites with information derived from PhosphositePlus.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.Dataframe</span>
<span class="sd">        dataframe containing PS of interst.</span>
<span class="sd">    ps : str</span>
<span class="sd">        Column containing info about the PS.</span>
<span class="sd">        Format: GeneName_AminoacidPositoin (e.g. AKT_T308).</span>
<span class="sd">    cols_to_keep : list, optional</span>
<span class="sd">        Which columns from original dataframe (input df) to keep in output.</span>
<span class="sd">        The default is None.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.Dataframe</span>
<span class="sd">        The input dataframe with the kept columns and additional phosphosite cols.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">cols_to_keep</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cols_to_keep</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">make_merge_col</span><span class="p">(</span><span class="n">df_to_merge</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">&quot;regSites&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Format the phosphosite positions and gene names so that merging is possible.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">file</span> <span class="o">==</span> <span class="s2">&quot;regSites&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">df_to_merge</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> \
                   <span class="n">df_to_merge</span><span class="p">[</span><span class="s2">&quot;MOD_RSD&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                       <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">df_to_merge</span><span class="p">[</span><span class="s2">&quot;SUB_GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> \
               <span class="n">df_to_merge</span><span class="p">[</span><span class="s2">&quot;SUB_MOD_RSD&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">resources</span><span class="o">.</span><span class="n">open_binary</span><span class="p">(</span><span class="s2">&quot;autoprot.data&quot;</span><span class="p">,</span> <span class="s2">&quot;Kinase_Substrate_Dataset.zip&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">d</span><span class="p">:</span>
        <span class="n">ks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;zip&#39;</span><span class="p">)</span>
        <span class="n">ks</span><span class="p">[</span><span class="s2">&quot;merge&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">make_merge_col</span><span class="p">(</span><span class="n">ks</span><span class="p">,</span> <span class="s2">&quot;KS&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">resources</span><span class="o">.</span><span class="n">open_binary</span><span class="p">(</span><span class="s2">&quot;autoprot.data&quot;</span><span class="p">,</span> <span class="s2">&quot;Regulatory_sites.zip&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">d</span><span class="p">:</span>
        <span class="n">reg_sites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;zip&#39;</span><span class="p">)</span>
        <span class="n">reg_sites</span><span class="p">[</span><span class="s2">&quot;merge&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">make_merge_col</span><span class="p">(</span><span class="n">reg_sites</span><span class="p">)</span>

    <span class="n">ks_coi</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KINASE&#39;</span><span class="p">,</span> <span class="s1">&#39;DOMAIN&#39;</span><span class="p">,</span> <span class="s1">&#39;IN_VIVO_RXN&#39;</span><span class="p">,</span> <span class="s1">&#39;IN_VITRO_RXN&#39;</span><span class="p">,</span> <span class="s1">&#39;CST_CAT#&#39;</span><span class="p">,</span> <span class="s1">&#39;merge&#39;</span><span class="p">]</span>
    <span class="n">reg_sites_coi</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ON_FUNCTION&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_OTHER_INTERACT&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;PMIDs&#39;</span><span class="p">,</span> <span class="s1">&#39;NOTES&#39;</span><span class="p">,</span> <span class="s1">&#39;LT_LIT&#39;</span><span class="p">,</span> <span class="s1">&#39;MS_LIT&#39;</span><span class="p">,</span> <span class="s1">&#39;MS_CST&#39;</span><span class="p">,</span> <span class="s1">&#39;merge&#39;</span><span class="p">]</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">ps</span><span class="p">:</span> <span class="s2">&quot;merge&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;merge&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cols_to_keep</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ks</span><span class="p">[</span><span class="n">ks_coi</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;merge&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">reg_sites</span><span class="p">[</span><span class="n">reg_sites_coi</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;merge&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="go_analysis"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.go_analysis">[docs]</a><span class="k">def</span> <span class="nf">go_analysis</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;hsapiens&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform go Enrichment analysis (also KEGG and REAC).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list of str</span>
<span class="sd">        list of gene names.</span>
<span class="sd">    organism : str, optional</span>
<span class="sd">        identifier for the organism.</span>
<span class="sd">        See https://biit.cs.ut.ee/gprofiler/page/organism-list for details.</span>
<span class="sd">        The default is &quot;hsapiens&quot;.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the input could not be parsed as list of gene names.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gp.profile</span>
<span class="sd">        Dataframe-like object with the GO annotations.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; autoprot.analysis.go_analysis([&#39;PEX14&#39;, &#39;PEX18&#39;]).iloc[:3,:3]</span>
<span class="sd">    source      native                                   name</span>
<span class="sd">    0  CORUM  CORUM:1984                 PEX14 homodimer complex</span>
<span class="sd">    1  GO:CC  GO:1990429          peroxisomal importomer complex</span>
<span class="sd">    2  GO:BP  GO:0036250  peroxisome transport along microtubule</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">gene_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Please provide a list of gene names&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gp</span><span class="o">.</span><span class="n">profile</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span> <span class="n">query</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">no_evidences</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_psm"><a class="viewcode-back" href="../../analysis.html#autoprot.analysis.make_psm">[docs]</a><span class="k">def</span> <span class="nf">make_psm</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq_len</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a position score matrix for a set of sequences.</span>

<span class="sd">    Returns the percentage of each amino acid for each position that</span>
<span class="sd">    can be further normalized using a PSM of unrelated/background sequences.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : list of str</span>
<span class="sd">        list of sequences.</span>
<span class="sd">    seq_len : int</span>
<span class="sd">        Length of the peptide sequences.</span>
<span class="sd">        Must match to the list provided.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.Dataframe</span>
<span class="sd">        Dataframe holding the prevalence for every amino acid per position in</span>
<span class="sd">        the input sequences.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; autoprot.analysis.make_psm([&#39;PEPTIDE&#39;, &#39;PEGTIDE&#39;, &#39;GGGGGGG&#39;], 7)</span>
<span class="sd">              0         1         2         3         4         5         6</span>
<span class="sd">    G  0.333333  0.333333  0.666667  0.333333  0.333333  0.333333  0.333333</span>
<span class="sd">    P  0.666667  0.000000  0.333333  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    matrix_a  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    V  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    L  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    I  0.000000  0.000000  0.000000  0.000000  0.666667  0.000000  0.000000</span>
<span class="sd">    M  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    C  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    F  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    Y  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    W  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    H  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    K  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    R  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    Q  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    N  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    E  0.000000  0.666667  0.000000  0.000000  0.000000  0.000000  0.666667</span>
<span class="sd">    D  0.000000  0.000000  0.000000  0.000000  0.000000  0.666667  0.000000</span>
<span class="sd">    S  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="sd">    T  0.000000  0.000000  0.000000  0.666667  0.000000  0.000000  0.000000</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">aa_dic</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;P&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;matrix_a&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;V&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;L&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;I&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;M&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;F&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;Y&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;W&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;H&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;K&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;R&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;Q&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;N&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;E&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;D&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;S&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">==</span> <span class="n">seq_len</span><span class="p">]</span>
    <span class="n">seq_t</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">seq</span><span class="p">)]</span>
    <span class="n">score_matrix</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">seq_t</span><span class="p">:</span>
        <span class="n">d</span> <span class="o">=</span> <span class="n">aa_dic</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">pos</span><span class="p">:</span>
            <span class="n">aa</span> <span class="o">=</span> <span class="n">aa</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s2">&quot;dataframe&quot;</span><span class="p">]:</span>
                <span class="n">d</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">score_matrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">score_matrix</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">pos</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">pos</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">/=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

    <span class="c1"># empty array -&gt; (sequenceWindow, aa)</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">seq_len</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">score_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="n">m</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span>

    <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">aa_dic</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">m</span><span class="o">.</span><span class="n">T</span></div>
</pre></div>

            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
</div>

<div class="toc-item">
  
</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/pydata-sphinx-theme.js?digest=9b1a4fa89bdd0e95b23b"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2022, The autoprot contributors.<br>

</p>

  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.0.2.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>