
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Analysis &#8212; autoprot  documentation</title>
  <script>
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=9b1a4fa89bdd0e95b23b" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=9b1a4fa89bdd0e95b23b" rel="stylesheet">

  
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/6.1.2/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/css/custom.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=9b1a4fa89bdd0e95b23b">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'analysis';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Clustering" href="clustering.html" />
    <link rel="prev" title="Data Preprocessing" href="preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fas fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="index.html">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">autoprot  documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                <li class="nav-item">
                    <a class="nav-link" href="installation.html">
                        Installation
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="preprocessing.html">
                        Data Preprocessing
                    </a>
                </li>
                

                <li class="nav-item current active">
                    <a class="nav-link" href="#">
                        Analysis
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="clustering.html">
                        Clustering
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="visualization.html">
                        Data Visualization
                    </a>
                </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                <li class="nav-item">
                    <a class="nav-link" href="workflows.html">
                        Workflows
                    </a>
                </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      <div class="navbar-end-item navbar-end__search-button-container">
        
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search">
  <i class="fas fa-search"></i>
</button>
      </div>
      
      <div class="navbar-end-item">
        <span class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle">
    <a class="theme-switch" data-mode="light"><i class="fas fa-sun"></i></a>
    <a class="theme-switch" data-mode="dark"><i class="far fa-moon"></i></a>
    <a class="theme-switch" data-mode="auto"><i class="fas fa-adjust"></i></a>
</span>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
  </div>


  
  <div class="search-button-container--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search">
  <i class="fas fa-search"></i>
</button>
  </div>

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fas fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                <li class="nav-item">
                    <a class="nav-link" href="installation.html">
                        Installation
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="preprocessing.html">
                        Data Preprocessing
                    </a>
                </li>
                

                <li class="nav-item current active">
                    <a class="nav-link" href="#">
                        Analysis
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="clustering.html">
                        Clustering
                    </a>
                </li>
                

                <li class="nav-item">
                    <a class="nav-link" href="visualization.html">
                        Data Visualization
                    </a>
                </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                <li class="nav-item">
                    <a class="nav-link" href="workflows.html">
                        Workflows
                    </a>
                </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <span class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle">
    <a class="theme-switch" data-mode="light"><i class="fas fa-sun"></i></a>
    <a class="theme-switch" data-mode="dark"><i class="far fa-moon"></i></a>
    <a class="theme-switch" data-mode="auto"><i class="fas fa-adjust"></i></a>
</span>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
      </ul>
      </div>
      
    </div>
    
  </div>

  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

      </div>
      <main class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="analysis">
<h1>Analysis<a class="headerlink" href="#analysis" title="Permalink to this heading">#</a></h1>
<p>The analysis submodule contains handy functions for the further analysis of your data. Functions include basic statistical tests, clustering and dimension reduction algorithms.</p>
<span class="target" id="module-autoprot.analysis"></span><p>Autoprot Analysis Functions.</p>
<p>&#64;author: Wignand</p>
<p>&#64;documentation: Julian</p>
<dl class="py class">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">AutoPCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlabels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clabels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA" title="Permalink to this definition">#</a></dt>
<dd><p>Conduct principal component analyses.</p>
<p>The class encompasses a set of helpful visualizations
for further investigating the results of the PCA
It needs the matrix on which the PCA is performed
as well as row labels (rlabels)
and column labels (clabels) corresponding to the
provided matrix.</p>
<p class="rubric">Notes</p>
<p>PCA is a method which allows you to visually investigate the underlying structure
in your data after reduction of the dimensionality.
With the .autoPCA() you can easily perform a PCA and also generate exploratory figures.
Intro to PCA: <a class="reference external" href="https://learnche.org/pid/latent-variable-modelling/principal-component-analysis/index">https://learnche.org/pid/latent-variable-modelling/principal-component-analysis/index</a></p>
<p class="rubric">Examples</p>
<p>for PCA no missing values are allowed
filter those and store complete dataframe</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">temp</span> <span class="o">=</span> <span class="n">prot</span><span class="p">[</span><span class="o">~</span><span class="n">prot</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;log2.*norm&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span>
</pre></div>
</div>
<p>get the matrix of quantitative values corresponding to conditions of interest
Here we only use the first replicate for clarity</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;log2.*norm.*_1$&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>generate appropiate names for the columns and rows of the matrix
for example here the columns represent the conditions and we are not interested in the rows (which are the genes)</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">clabels</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">columns</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rlabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
</pre></div>
</div>
<p>generate autopca object</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span> <span class="o">=</span> <span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">AutoPCA</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">rlabels</span><span class="p">,</span> <span class="n">clabels</span><span class="p">)</span>
</pre></div>
</div>
<p>The scree plots describe how much of the total variance of the dataset is
explained ba the first n components. As you want to explain as variance as
possible with as little variables as possible, chosing the number of components
directly right to the steep descend of the plot is usually a good idea.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">scree</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-1.py">Source code</a>, <a class="reference external" href=".//analysis-1.png">png</a>, <a class="reference external" href=".//analysis-1.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-1.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-1.png" class="plot-directive" src="_images/analysis-1.png" />
</figure>
<p>The corrComp heatmap shows the PCA loads (i.e. how much a principal component is
influenced by a change in that variable) relative to the variables (i.e. the
experiment conditions). If a weight (colorbar) is close to zero, the corresponding
PC is barely influenced by it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">corr_comp</span><span class="p">(</span><span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-2.py">Source code</a>, <a class="reference external" href=".//analysis-2.png">png</a>, <a class="reference external" href=".//analysis-2.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-2.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-2.png" class="plot-directive" src="_images/analysis-2.png" />
</figure>
<p>The bar loading plot is a different way to represent the weights/loads for each
condition and principal component. High values indicate a high influence of the
variable/condition on the PC.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">bar_load</span><span class="p">(</span><span class="n">pc</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">bar_load</span><span class="p">(</span><span class="n">pc</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-3.py">Source code</a>)</p>
<figure class="align-default" id="id1">
<img alt="_images/analysis-3_00.png" class="plot-directive" src="_images/analysis-3_00.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-3_00.png">png</a>, <a class="reference external" href=".//analysis-3_00.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-3_00.pdf">pdf</a>)</span><a class="headerlink" href="#id1" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="id2">
<img alt="_images/analysis-3_01.png" class="plot-directive" src="_images/analysis-3_01.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-3_01.png">png</a>, <a class="reference external" href=".//analysis-3_01.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-3_01.pdf">pdf</a>)</span><a class="headerlink" href="#id2" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The score plot shows how the different data points (i.e. proteins) are positioned
with respect to two principal components.
In more detail, the scores are the original data values multiplied by the
weights of each value for each principal component.
Usually they will separate more in the direction of PC1 as this component
explains the largest share of the data variance</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">score_plot</span><span class="p">(</span><span class="n">pc1</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pc2</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-4.py">Source code</a>, <a class="reference external" href=".//analysis-4.png">png</a>, <a class="reference external" href=".//analysis-4.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-4.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-4.png" class="plot-directive" src="_images/analysis-4.png" />
</figure>
<p>The loading plot is the 2D representation of the barLoading plots and shows
the weights how each variable influences the two PCs.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autopca</span><span class="o">.</span><span class="n">loading_plot</span><span class="p">(</span><span class="n">pc1</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pc2</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">labeling</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-5.py">Source code</a>, <a class="reference external" href=".//analysis-5.png">png</a>, <a class="reference external" href=".//analysis-5.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-5.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-5.png" class="plot-directive" src="_images/analysis-5.png" />
</figure>
<p>The Biplot is a combination of loading plot and score plot as it shows the
scores for each protein as point and the weights for each variable as
vectors.
&gt;&gt;&gt; autopca.bi_plot(pc1=1, pc2=2)</p>
<p>(<a class="reference external" href=".//analysis-6.py">Source code</a>, <a class="reference external" href=".//analysis-6.png">png</a>, <a class="reference external" href=".//analysis-6.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-6.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-6.png" class="plot-directive" src="_images/analysis-6.png" />
</figure>
<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.bar_load">
<span class="sig-name descname"><span class="pre">bar_load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.bar_load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.bar_load" title="Permalink to this definition">#</a></dt>
<dd><p>Plot the loadings of a given component in a barplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc</strong> (<em>int</em><em>, </em><em>optional</em>) – Component to draw. The default is 1.</p></li>
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Plot only the n first rows.
The default is 25.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.bi_plot">
<span class="sig-name descname"><span class="pre">bi_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_load</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.bi_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.bi_plot" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a biplot, a combined loadings and score plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc1</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the first PC to plot. The default is 1.</p></li>
<li><p><strong>pc2</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the second PC to plot. The default is 2.</p></li>
<li><p><strong>num_load</strong> (<em>'all'</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – Plot only the n first rows.
The default is “all”.</p></li>
<li><p><strong>figsize</strong> (<em>tuple of int</em><em>, </em><em>optional</em>) – Figure size. The default is (3,3).</p></li>
<li><p><strong>**kwargs</strong> – Passed to plt.scatter.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>In the biplot, scores are shown as points and loadings as
vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.corr_comp">
<span class="sig-name descname"><span class="pre">corr_comp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.corr_comp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.corr_comp" title="Permalink to this definition">#</a></dt>
<dd><p>Plot heatmap of PCA weights vs. variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>annot</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, write the data value in each cell.
If an array-like with the same shape as data, then use this
to annotate the heatmap instead of the data.
Note that DataFrames will match on position, not index.
The default is False.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>2D representation how strong each observation (e.g. log protein ratio)
weights for each principal component.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.loading_plot">
<span class="sig-name descname"><span class="pre">loading_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labeling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.loading_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.loading_plot" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a PCA loading plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc1</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the first PC to plot. The default is 1.</p></li>
<li><p><strong>pc2</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the second PC to plot. The default is 2.</p></li>
<li><p><strong>labeling</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, points are labelled with the corresponding
column labels. The default is False.</p></li>
<li><p><strong>figsize</strong> (<em>tuple of int</em><em>, </em><em>optional</em>) – The size of the figure object.
The default is (5,5).</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This will return a scatterplot with as many points as there are
components (i.e. conditions) in the dataset.
For each component a load magnitude for two PCs will be printedd
that describes how much each condition influences the magnitude
of the respective PC.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.pair_plot">
<span class="sig-name descname"><span class="pre">pair_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.pair_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.pair_plot" title="Permalink to this definition">#</a></dt>
<dd><p>Draw a pair plot of for pca for the given number of dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Plot only the n first rows. The default is 0.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Be careful for large data this might crash you PC -&gt; better specify n!</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.return_load">
<span class="sig-name descname"><span class="pre">return_load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.return_load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.return_load" title="Permalink to this definition">#</a></dt>
<dd><p>Return the load for a given principal component.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc</strong> (<em>int</em><em>, </em><em>optional</em>) – Component to draw. The default is 1.</p></li>
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Plot only the n first rows.
The default is 25.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing load vs. condition.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.return_score">
<span class="sig-name descname"><span class="pre">return_score</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.return_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.return_score" title="Permalink to this definition">#</a></dt>
<dd><p>Return a dataframe of all scorings for all principal components.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>scores</strong> – Dataframe holding the principal components as colnames and
the scores for each protein on that PC as values.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.score_plot">
<span class="sig-name descname"><span class="pre">score_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pc1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labeling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.score_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.score_plot" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a PCA score plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pc1</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the first PC to plot. The default is 1.</p></li>
<li><p><strong>pc2</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of the second PC to plot. The default is 2.</p></li>
<li><p><strong>labeling</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, points are labelled with the corresponding
column labels. The default is False.</p></li>
<li><p><strong>file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save the plot. The default is None.</p></li>
<li><p><strong>figsize</strong> (<em>tuple of int</em><em>, </em><em>optional</em>) – Figure size. The default is (5,5).</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This will return a scatterplot with as many points as there are
entries (i.e. protein IDs).
The scores for each PC are the original protein ratios multiplied with
the loading weights.
The score plot corresponds to the individual positions of of each protein
on a hyperplane generated by the pc1 and pc2 vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.AutoPCA.scree">
<span class="sig-name descname"><span class="pre">scree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#AutoPCA.scree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.AutoPCA.scree" title="Permalink to this definition">#</a></dt>
<dd><p>Plot Scree plot and Explained variance vs number of components.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>figsize</strong> (<em>tuple of int</em><em>, </em><em>optional</em>) – The size of the figure object.
The default is (15,5).</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – No PCA object was initialised in the class.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">KSEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA" title="Permalink to this definition">#</a></dt>
<dd><p>Perform kinase substrate enrichment analysis.</p>
<p class="rubric">Notes</p>
<p>KSEA uses the Kinase-substrate dataset and the
regulatory-sites dataset from <a class="reference external" href="https://www.phosphosite.org/staticDownloads">https://www.phosphosite.org/staticDownloads</a></p>
<p class="rubric">Examples</p>
<p>KSEA is a method to get insights on which kinases are active in a given
phosphoproteomic dataset. This is a great method to gain deeper insights
on the underlying signaling mechanisms and also to generate novel
hypothesis and find new connections in signaling processes.
The KSEA class allows you to easily perform the analysis and
comes with helpful functions to visualize and interpret your results.</p>
<p>In the first step of the analysis you have to generate a KSEA object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span> <span class="o">=</span> <span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">KSEA</span><span class="p">(</span><span class="n">phos</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, you can annotate the data with respective kinases.
You can provide the function with a organism of your choice as well as
toggle whether or not to screen for only in vivo determined substrate
phosphorylation of the respective kinases.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;mouse&quot;</span><span class="p">,</span> <span class="n">only_in_vivo</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>After the annotation it is always a good idea to get an overview of the
kinases in the data an how many substrates the have. Based on this you
might want to adjust a cutoff specifying the minimum number of substrates
per kinase.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">get_kinase_overview</span><span class="p">(</span><span class="n">kois</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Akt1&quot;</span><span class="p">,</span><span class="s2">&quot;MKK4&quot;</span><span class="p">,</span> <span class="s2">&quot;P38A&quot;</span><span class="p">,</span> <span class="s2">&quot;Erk1&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Next, you can perform the actual kinase substrate enrichment analysis.
The analysis is based on the log fold change of your data.
Therefore, you have to provide the function with the appropiate column of
your data and the minimum number of substrates per kinase.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">ksea</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="s2">&quot;logFC_TvC&quot;</span><span class="p">,</span> <span class="n">min_subs</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<p>After the ksea has finished, you can get information for further analysis
such as the substrates of a specific kinase (or a list of kinases)</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">return_kinase_substrate</span><span class="p">(</span><span class="n">kinase</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Akt1&quot;</span><span class="p">,</span> <span class="s2">&quot;MKK4&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">sample</span><span class="p">()</span> 
</pre></div>
</div>
<p>or a new dataframe with additional columns for every kinase showing if the
protein is a substrate of that kinase or not</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">annotate_df</span><span class="p">(</span><span class="n">kinases</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Akt1&quot;</span><span class="p">,</span> <span class="s2">&quot;MKK4&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="p">:]</span>
</pre></div>
</div>
<p>Eventually, you can also generate plots of the enrichment analysis.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">plot_enrichment</span><span class="p">(</span><span class="n">up_col</span><span class="o">=</span><span class="s2">&quot;salmon&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-7.py">Source code</a>)</p>
<figure class="align-default" id="id3">
<img alt="_images/analysis-7_00.png" class="plot-directive" src="_images/analysis-7_00.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-7_00.png">png</a>, <a class="reference external" href=".//analysis-7_00.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-7_00.pdf">pdf</a>)</span><a class="headerlink" href="#id3" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="id4">
<img alt="_images/analysis-7_01.png" class="plot-directive" src="_images/analysis-7_01.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-7_01.png">png</a>, <a class="reference external" href=".//analysis-7_01.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-7_01.pdf">pdf</a>)</span><a class="headerlink" href="#id4" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>You can also highlight a list of kinases in volcano plots.
This is based on the autoprot volcano function.
You can pass all the common parameters to this function.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">volcanos</span><span class="p">(</span><span class="n">log_fc</span><span class="o">=</span><span class="s2">&quot;logFC_TvC&quot;</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="s2">&quot;pValue_TvC&quot;</span><span class="p">,</span> <span class="n">kinases</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Akt1&quot;</span><span class="p">,</span> <span class="s2">&quot;MKK4&quot;</span><span class="p">],</span>
<span class="gp">... </span>              <span class="n">annot</span><span class="o">=</span><span class="s2">&quot;Gene names&quot;</span><span class="p">,</span> <span class="n">sig_col</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-8.py">Source code</a>)</p>
<figure class="align-default" id="id5">
<img alt="_images/analysis-8_00.png" class="plot-directive" src="_images/analysis-8_00.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-8_00.png">png</a>, <a class="reference external" href=".//analysis-8_00.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-8_00.pdf">pdf</a>)</span><a class="headerlink" href="#id5" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="id6">
<img alt="_images/analysis-8_01.png" class="plot-directive" src="_images/analysis-8_01.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-8_01.png">png</a>, <a class="reference external" href=".//analysis-8_01.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-8_01.pdf">pdf</a>)</span><a class="headerlink" href="#id6" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Sometimes the enrichment is crowded by various kinase isoforms.
In such cases it makes sense to simplify the annotation by grouping those
isoforms together.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">simplify</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ERK&quot;</span><span class="p">:[</span><span class="s2">&quot;ERK1&quot;</span><span class="p">,</span><span class="s2">&quot;ERK2&quot;</span><span class="p">],</span>
<span class="gp">... </span>            <span class="s2">&quot;GSK3&quot;</span><span class="p">:[</span><span class="s2">&quot;GSK3A&quot;</span><span class="p">,</span> <span class="s2">&quot;GSK3B&quot;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">ksea</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="s2">&quot;logFC_TvC&quot;</span><span class="p">,</span> <span class="n">min_subs</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">simplify</span><span class="o">=</span><span class="n">simplify</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">plot_enrichment</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-9.py">Source code</a>, <a class="reference external" href=".//analysis-9.png">png</a>, <a class="reference external" href=".//analysis-9.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-9.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-9.png" class="plot-directive" src="_images/analysis-9.png" />
</figure>
<p>Of course, you can also get the ksea results as a dataframe to save or to further customize.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">return_enrichment</span><span class="p">()</span>
</pre></div>
</div>
<p>Of course is the database not exhaustive and you might want to add additional
substrates manually. This can be done the following way.
Manually added substrates are always added irrespective of the species used
for the annotation.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span> <span class="o">=</span> <span class="n">ana</span><span class="o">.</span><span class="n">KSEA</span><span class="p">(</span><span class="n">phos</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;RPGR&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">modRsds</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;S564&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kinases</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;mTOR&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">add_substrate</span><span class="p">(</span><span class="n">kinase</span><span class="o">=</span><span class="n">kinases</span><span class="p">,</span> <span class="n">substrate</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">sub_mod_rsd</span><span class="o">=</span><span class="n">modRsds</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;mouse&quot;</span><span class="p">,</span> <span class="n">only_in_vivo</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">ksea</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="s2">&quot;logFC_TvC&quot;</span><span class="p">,</span> <span class="n">min_subs</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">plot_enrichment</span><span class="p">(</span><span class="n">plot_bg</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">clear_manual_substrates</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s2">&quot;mouse&quot;</span><span class="p">,</span> <span class="n">only_in_vivo</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">ksea</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="s2">&quot;logFC_TvC&quot;</span><span class="p">,</span> <span class="n">min_subs</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ksea</span><span class="o">.</span><span class="n">plot_enrichment</span><span class="p">(</span><span class="n">plot_bg</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.add_substrate">
<span class="sig-name descname"><span class="pre">add_substrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kinase</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substrate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sub_mod_rsd</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.add_substrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.add_substrate" title="Permalink to this definition">#</a></dt>
<dd><p>Manually add a substrate to the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kinase</strong> (<em>list of str</em>) – Name of the kinase e.g. PAK2.</p></li>
<li><p><strong>substrate</strong> (<em>list of str</em>) – Name of the substrate e.g. Prkd1.</p></li>
<li><p><strong>sub_mod_rsd</strong> (<em>list of str</em>) – Phosphorylated residues e.g. S203.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the three provided lists do not match in length.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.annotate">
<span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_in_vivo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.annotate" title="Permalink to this definition">#</a></dt>
<dd><p>Annotate with known kinase substrate pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism</strong> (<em>str</em><em>, </em><em>optional</em>) – The target organism. The default is “human”.</p></li>
<li><p><strong>only_in_vivo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to restrict analysis to in vivo evidence.
The default is False.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Manually added kinases will be included in the annotation search
independent of the setting of organism and onInVivo.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.annotate_df">
<span class="sig-name descname"><span class="pre">annotate_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kinases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.annotate_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.annotate_df" title="Permalink to this definition">#</a></dt>
<dd><p>Annotate the provided dataframe with boolean columns for given kinases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>kinases</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of kinases. The default is [].</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>annotated dataframe containing a column for each provided kinase
with boolean values representing a row/protein being a kinase
substrate or not.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.clear_manual_substrates">
<span class="sig-name descname"><span class="pre">clear_manual_substrates</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.clear_manual_substrates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.clear_manual_substrates" title="Permalink to this definition">#</a></dt>
<dd><p>Remove all manual entries from the PSP database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.get_kinase_overview">
<span class="sig-name descname"><span class="pre">get_kinase_overview</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.get_kinase_overview"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.get_kinase_overview" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a graphical overview of the kinases acting on the proteins in the dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>kois</strong> (<em>list of str</em><em>, </em><em>optional</em>) – Kinases of interest for which a detailed overview of substrate numbers
is plotted. The default is None.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.ksea">
<span class="sig-name descname"><span class="pre">ksea</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_subs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simplify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.ksea"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.ksea" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate Kinase Enrichment Score.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>str</em>) – Column used for the analysis containing the kinase substrate
enrichments.</p></li>
<li><p><strong>min_subs</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of substrates a kinase must have to be considered.
The default is 5.</p></li>
<li><p><strong>simplify</strong> (<em>None</em><em>, </em><em>&quot;auto&quot;</em><em> or </em><em>dict</em><em>, </em><em>optional</em>) – Merge multiple kinases during analysis.
Using “auto” a predefined set of kinase isoforms is merged.
If provided with a dict, the dict has to contain a list of kinases
to merge as values and the name of the merged kinases as key.
The default is None.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The enrichment score is calculated as</p>
<div class="math notranslate nohighlight">
\[\frac{(\langle FC_{kinase} \rangle - \langle FC_{all} \rangle)\sqrt{N_{all}}}{\sigma_{all}}\]</div>
<p>i.e. the difference in mean fold change between kinase and all substrates
multiplied by the square root of number of kinase substrates and divided
by the standard deviation of the fold change of all substrates.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.plot_enrichment">
<span class="sig-name descname"><span class="pre">plot_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">up_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'orange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">down_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bg_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lightgray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_bg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ret</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.plot_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.plot_enrichment" title="Permalink to this definition">#</a></dt>
<dd><p>Plot the KSEA results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>up_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for enriched/upregulated kinases.
The default is “orange”.</p></li>
<li><p><strong>down_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Colour for deriched/downregulated kinases.
The default is “blue”.</p></li>
<li><p><strong>bg_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Colour for not kinases that did not change significantly.
The default is “lightgray”.</p></li>
<li><p><strong>plot_bg</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to plot the unaffected kinases.
The default is True.</p></li>
<li><p><strong>ret</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return the figure object.
The default is False.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – Title of the figure. The default is “”.</p></li>
<li><p><strong>figsize</strong> (<em>tuple of int</em><em>, </em><em>optional</em>) – Figure size. The default is (5,10).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fig</strong> – Only returned in ret is True.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib figure.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.return_enrichment">
<span class="sig-name descname"><span class="pre">return_enrichment</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.return_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.return_enrichment" title="Permalink to this definition">#</a></dt>
<dd><p>Return a dataframe of kinase:score pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.return_kinase_substrate">
<span class="sig-name descname"><span class="pre">return_kinase_substrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kinase</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.return_kinase_substrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.return_kinase_substrate" title="Permalink to this definition">#</a></dt>
<dd><p>Return new dataframe with substrates of one or multiple kinase(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>kinase</strong> (<em>str</em><em> or </em><em>list of str</em>) – Kinase(s) to analyse.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If kinase is neither list of str nor str.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df_filter</strong> – Dataframe containing detailed information on kinase-substrate pairs
including reference literature.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.Dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="autoprot.analysis.KSEA.volcanos">
<span class="sig-name descname"><span class="pre">volcanos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">log_fc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#KSEA.volcanos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.KSEA.volcanos" title="Permalink to this definition">#</a></dt>
<dd><p>Plot volcano plots highlighting substrates of a given kinase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>log_fc</strong> (<em>str</em>) – Column name of column containing the log fold changes.
Must be present in the dataframe KSEA was initialised with.</p></li>
<li><p><strong>p</strong> (<em>str</em>) – Column name of column containing the p values.
Must be present in the dataframe KSEA was initialised with.</p></li>
<li><p><strong>kinases</strong> (<em>list of str</em><em>, </em><em>optional</em>) – Limit the analysis to these kinases. The default is [].</p></li>
<li><p><strong>**kwargs</strong> – passed to autoprot.visualisation.volcano.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.adjust_p">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">adjust_p</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdr_bh'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#adjust_p"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.adjust_p" title="Permalink to this definition">#</a></dt>
<dd><p>Use statsmodels.multitest on dataframes.</p>
<p>Note: when nan in p-value this function will return only nan.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input dataframe.</p></li>
<li><p><strong>p_col</strong> (<em>str</em>) – column containing p-values for correction.</p></li>
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em>) – ‘b’: ‘Bonferroni’,
‘s’: ‘Sidak’,
‘h’: ‘Holm’,
‘hs’: ‘Holm-Sidak’,
‘sh’: ‘Simes-Hochberg’,
‘ho’: ‘Hommel’,
‘fdr_bh’: ‘FDR Benjamini-Hochberg’,
‘fdr_by’: ‘FDR Benjamini-Yekutieli’,
‘fdr_tsbh’: ‘FDR 2-stage Benjamini-Hochberg’,
‘fdr_tsbky’: ‘FDR 2-stage Benjamini-Krieger-Yekutieli’,
‘fdr_gbs’: ‘FDR adaptive Gavrilov-Benjamini-Sarkar’
The default is “fdr_bh”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – The input dataframe with adjusted p-values.
The dataframe will have a column named “adj.{pCol}”</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">twitchVsmild</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;log2_Ratio H/M normalized BC18_1&#39;</span><span class="p">,</span><span class="s1">&#39;log2_Ratio M/L normalized BC18_2&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio H/M normalized BC18_3&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio H/L normalized BC36_1&#39;</span><span class="p">,</span><span class="s1">&#39;log2_Ratio H/M normalized BC36_2&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio M/L normalized BC36_2&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;_static/testdata/proteinGroups.zip&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protRatio</span> <span class="o">=</span> <span class="n">prot</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;Ratio .\/. normalized&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protLog</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">prot</span><span class="p">,</span> <span class="n">protRatio</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_tt</span> <span class="o">=</span> <span class="n">ana</span><span class="o">.</span><span class="n">ttest</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">protLog</span><span class="p">,</span> <span class="n">reps</span><span class="o">=</span><span class="n">twitchVsmild</span><span class="p">,</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;TvM&quot;</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">adjust_p_vals</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_tt_adj</span> <span class="o">=</span> <span class="n">ana</span><span class="o">.</span><span class="n">adjust_p</span><span class="p">(</span><span class="n">prot_tt</span><span class="p">,</span> <span class="n">p_col</span><span class="o">=</span><span class="s2">&quot;pValue_TvM&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_tt_adj</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s1">&#39;pValue&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">   pValue_TvM  adj.pValue_TvM</span>
<span class="go">0         NaN             NaN</span>
<span class="go">1    0.947334        0.966514</span>
<span class="go">2         NaN             NaN</span>
<span class="go">3         NaN             NaN</span>
<span class="go">4    0.031292        0.206977</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.annotate_phosphosite">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">annotate_phosphosite</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols_to_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#annotate_phosphosite"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.annotate_phosphosite" title="Permalink to this definition">#</a></dt>
<dd><p>Annotate phosphosites with information derived from PhosphositePlus.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.Dataframe</em>) – dataframe containing PS of interst.</p></li>
<li><p><strong>ps</strong> (<em>str</em>) – Column containing info about the PS.
Format: GeneName_AminoacidPositoin (e.g. AKT_T308).</p></li>
<li><p><strong>cols_to_keep</strong> (<em>list</em><em>, </em><em>optional</em>) – Which columns from original dataframe (input df) to keep in output.
The default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The input dataframe with the kept columns and additional phosphosite cols.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.cohen_d">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">cohen_d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#cohen_d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.cohen_d" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate Cohen’s d effect size for two groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.Dataframe</em>) – Input dataframe.</p></li>
<li><p><strong>group1</strong> (<em>str</em>) – Colname for group1.</p></li>
<li><p><strong>group2</strong> (<em>str</em>) – Colname for group2.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – Input dataframe with a new column “cohenD”.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Dataframe</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Cohen’s d is defined as the difference between two means divided by a standard deviation for the data.
Note that the pooled standard deviation here is calculated differently than
originally proposed by Cohen.</p>
<p class="rubric">References</p>
<p>[1] Cohen, Jacob (1988). Statistical Power Analysis for the Behavioral Sciences. Routledge.
[2] <a class="reference external" href="https://www.doi.org/10.22237/jmasm/1257035100">https://www.doi.org/10.22237/jmasm/1257035100</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.edm">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">edm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#edm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.edm" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate an euclidean distance matrix between two matrices.</p>
<p>See:  <a class="reference external" href="https://medium.com/swlh/euclidean-distance-matrix-4c3e1378d87f">https://medium.com/swlh/euclidean-distance-matrix-4c3e1378d87f</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix_a</strong> (<em>np.ndarray</em>) – Matrix 1.</p></li>
<li><p><strong>matrix_b</strong> (<em>np.ndarray</em>) – Matrix 2.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Distance matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.get_pub_abstracts">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">get_pub_abstracts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title_or_abstract</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">author</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phrase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclusion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_search_term</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">make_word_cloud</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pub+date'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'print'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#get_pub_abstracts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.get_pub_abstracts" title="Permalink to this definition">#</a></dt>
<dd><p>Get Pubmed abstracts.</p>
<p class="rubric">Notes</p>
<p>Provide the search terms and pubmed will be searched for paper matching those
The exclusion list can be used to exclude certain words to be contained in a paper
If you want to perform more complicated searches use the possibility of using
a custom term.</p>
<p>If you search for a not excact term you can use *
For example opto* will also return optogenetic.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>text</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of words which must occur in text.
The default is None.</p></li>
<li><p><strong>title_or_abstract</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of words which must occur in Titel or Abstract.
The default is None.</p></li>
<li><p><strong>author</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of authors which must occor in Author list.
The default is None.</p></li>
<li><p><strong>phrase</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of phrases which should occur in text.
Seperate word in phrases with hyphen (e.g. akt-signaling).
The default is None.</p></li>
<li><p><strong>exclusion</strong> (<em>list of str</em><em>, </em><em>optional</em>) – List of words which must not occur in text.
The default is None.</p></li>
<li><p><strong>custom_search_term</strong> (<em>str</em><em>, </em><em>optional</em>) – Enter a custom search term (make use of advanced pubmed search functionality).
If this is supplied all other search terms will be ignored.
The default is None.</p></li>
<li><p><strong>make_word_cloud</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to draw a wordcloud of the words in retrieved abstrace.
The default is False.</p></li>
<li><p><strong>sort</strong> (<em>str</em><em>, </em><em>optional</em>) – How to sort the results.
The default is ‘pub+date’.</p></li>
<li><p><strong>retmax</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of found articles to return.
The default is 20.</p></li>
<li><p><strong>output</strong> (<em>str</em><em>, </em><em>optional</em>) – How to handle the output.
Possible values are ‘SOMEPATH.txt’,
‘SOMEPATH.html’ or ‘SOMEPATH’.
If no extension is given, the output will be html.
The default is “print”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To generate a wordcloud and print the results of the found articles to the
prompt use the following command</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">get_pub_abstracts</span><span class="p">(</span><span class="n">titel_or_abstract</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;p38&quot;</span><span class="p">,</span> <span class="s2">&quot;JNK&quot;</span><span class="p">,</span> <span class="s2">&quot;ERK&quot;</span><span class="p">],</span>
<span class="go">                                      make_word_cloud=True)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-10.py">Source code</a>, <a class="reference external" href=".//analysis-10.png">png</a>, <a class="reference external" href=".//analysis-10.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-10.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-10.png" class="plot-directive" src="_images/analysis-10.png" />
</figure>
<p>Even more comfortably, you can also save the results incl. the wordcloud
as html file</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span> <span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">get_pub_abstracts</span><span class="p">(</span><span class="n">titel_or_abstract</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;p38&quot;</span><span class="p">,</span> <span class="s2">&quot;JNK&quot;</span><span class="p">,</span> <span class="s2">&quot;ERK&quot;</span><span class="p">],</span>
<span class="go">                                      make_word_cloud=True,</span>
<span class="go">                                      output=&#39;./MyPubmedSearch.html&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.go_analysis">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">go_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hsapiens'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#go_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.go_analysis" title="Permalink to this definition">#</a></dt>
<dd><p>Perform go Enrichment analysis (also KEGG and REAC).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>list of str</em>) – list of gene names.</p></li>
<li><p><strong>organism</strong> (<em>str</em><em>, </em><em>optional</em>) – identifier for the organism.
See <a class="reference external" href="https://biit.cs.ut.ee/gprofiler/page/organism-list">https://biit.cs.ut.ee/gprofiler/page/organism-list</a> for details.
The default is “hsapiens”.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input could not be parsed as list of gene names.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dataframe-like object with the GO annotations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>gp.profile</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">go_analysis</span><span class="p">([</span><span class="s1">&#39;PEX14&#39;</span><span class="p">,</span> <span class="s1">&#39;PEX18&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">3</span><span class="p">,:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">source      native                                   name</span>
<span class="go">0  CORUM  CORUM:1984                 PEX14 homodimer complex</span>
<span class="go">1  GO:CC  GO:1990429          peroxisomal importomer complex</span>
<span class="go">2  GO:BP  GO:0036250  peroxisome transport along microtubule</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.limma">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">limma</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_design</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_contrasts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#limma"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.limma" title="Permalink to this definition">#</a></dt>
<dd><p>Perform moderated ttest as implemented from R LIMMA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input dataframe.</p></li>
<li><p><strong>reps</strong> (<em>list of lists of str</em>) – Column names of the replicates.
Common replicates are grouped together in a single list.</p></li>
<li><p><strong>cond</strong> (<em>str</em><em>, </em><em>optional</em>) – Term to append to the newly generated colnames.
The default is “”.</p></li>
<li><p><strong>custom_design</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to custom design file.
The default is None.</p></li>
<li><p><strong>calc_contrasts</strong> (<em>str</em><em>, </em><em>optional</em>) – The contrasts to be calculated by limma. Must refer to the column names
in the data file. Differences are indicated e.g. by “CondA-CondB”.
The autoprot implementation only calculates contrasts based on the
column names specified in the design matrix.
The default is None.</p></li>
<li><p><strong>print_r</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to print the R output.
The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – The input dataframe with additional columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>matrix_a custom design representing the design matrix of the microarray experiment,
with rows corresponding to arrays and columns to coefficients to be estimated
can be provided using customDesign.
If customDesign is the unit vector meaning that the arrays are treated as replicates.
See: <a class="reference external" href="https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/lmFit">https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/lmFit</a></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;a1&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="s2">&quot;a2&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="s2">&quot;a3&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="s2">&quot;b1&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="s2">&quot;b2&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                     <span class="s2">&quot;b3&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">testRes</span> <span class="o">=</span> <span class="n">ana</span><span class="o">.</span><span class="n">limma</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">reps</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;a1&quot;</span><span class="p">,</span><span class="s2">&quot;a2&quot;</span><span class="p">,</span> <span class="s2">&quot;a3&quot;</span><span class="p">],[</span><span class="s2">&quot;b1&quot;</span><span class="p">,</span><span class="s2">&quot;b2&quot;</span><span class="p">,</span> <span class="s2">&quot;b3&quot;</span><span class="p">]],</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;_test&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">testRes</span><span class="p">[</span><span class="s2">&quot;P.Value_test&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-11.py">Source code</a>, <a class="reference external" href=".//analysis-11.png">png</a>, <a class="reference external" href=".//analysis-11.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-11.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-11.png" class="plot-directive" src="_images/analysis-11.png" />
</figure>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.loess">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">loess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xvals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">yvals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poly_degree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#loess"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.loess" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate a LOcally-Weighted Scatterplot Smoothing Fit.</p>
<p>See: <a class="reference external" href="https://medium.com/&#64;langen.mu/creating-powerfull-lowess-graphs-in-python-e0ea7a30b17a">https://medium.com/&#64;langen.mu/creating-powerfull-lowess-graphs-in-python-e0ea7a30b17a</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.Dataframe</em>) – Input dataframe.</p></li>
<li><p><strong>xvals</strong> (<em>str</em>) – Colname of x values.</p></li>
<li><p><strong>yvals</strong> (<em>str</em>) – Colname of y values.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Sensitivity of the estimation.
Controls how much of the total number of values is used during weighing.
0 &lt;= alpha &lt;= 1.</p></li>
<li><p><strong>poly_degree</strong> (<em>int</em><em>, </em><em>optional</em>) – Degree of the fitted polynomial. The default is 2.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Loess normalisation (also referred to as Savitzky-Golay filter) locally approximates
the data around every point using low-order functions and giving less weight to distant
data points.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="o">-</span><span class="mi">50</span><span class="p">,</span><span class="mi">110</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">x_values</span><span class="p">)</span><span class="o">/</span><span class="mf">1.5</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="o">-</span><span class="mi">1000</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">x_values</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Xvalue&quot;</span> <span class="p">:</span> <span class="n">x_values</span><span class="p">,</span>
<span class="go">                       &quot;Yvalue&quot; : y_values</span>
<span class="go">                       })</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">evalDF</span> <span class="o">=</span> <span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">loess</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;Xvalue&quot;</span><span class="p">,</span> <span class="s2">&quot;Yvalue&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">poly_degree</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Xvalue&quot;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Yvalue&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">eval_df</span><span class="p">[</span><span class="s1">&#39;v&#39;</span><span class="p">],</span> <span class="n">eval_df</span><span class="p">[</span><span class="s1">&#39;g&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Test&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-12.py">Source code</a>, <a class="reference external" href=".//analysis-12.png">png</a>, <a class="reference external" href=".//analysis-12.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-12.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-12.png" class="plot-directive" src="_images/analysis-12.png" />
</figure>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.make_psm">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">make_psm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#make_psm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.make_psm" title="Permalink to this definition">#</a></dt>
<dd><p>Generate a position score matrix for a set of sequences.</p>
<p>Returns the percentage of each amino acid for each position that
can be further normalized using a PSM of unrelated/background sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>list of str</em>) – list of sequences.</p></li>
<li><p><strong>seq_len</strong> (<em>int</em>) – Length of the peptide sequences.
Must match to the list provided.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe holding the prevalence for every amino acid per position in
the input sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Dataframe</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">make_psm</span><span class="p">([</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">,</span> <span class="s1">&#39;PEGTIDE&#39;</span><span class="p">,</span> <span class="s1">&#39;GGGGGGG&#39;</span><span class="p">],</span> <span class="mi">7</span><span class="p">)</span>
<span class="go">          0         1         2         3         4         5         6</span>
<span class="go">G  0.333333  0.333333  0.666667  0.333333  0.333333  0.333333  0.333333</span>
<span class="go">P  0.666667  0.000000  0.333333  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">matrix_a  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">V  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">L  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">I  0.000000  0.000000  0.000000  0.000000  0.666667  0.000000  0.000000</span>
<span class="go">M  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">C  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">F  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">Y  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">W  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">H  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">K  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">R  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">Q  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">N  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">E  0.000000  0.666667  0.000000  0.000000  0.000000  0.000000  0.666667</span>
<span class="go">D  0.000000  0.000000  0.000000  0.000000  0.000000  0.666667  0.000000</span>
<span class="go">S  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000</span>
<span class="go">T  0.000000  0.000000  0.000000  0.666667  0.000000  0.000000  0.000000</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.miss_analysis">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">miss_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ascending'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_vis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#miss_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.miss_analysis" title="Permalink to this definition">#</a></dt>
<dd><p>Print missing statistics for a dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.Dataframe</em>) – Input dataframe with missing values.</p></li>
<li><p><strong>cols</strong> (<em>list of str</em>) – Columns to perform missing values analysis on.</p></li>
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – How many rows of the dataframe to displayed.
The default is None (uses all rows).</p></li>
<li><p><strong>sort</strong> (<em>str</em><em>, </em><em>optional</em>) – “ascending” or “descending”.
The default is ‘ascending’.</p></li>
<li><p><strong>text</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to output text summaryMap.
The default is True.</p></li>
<li><p><strong>vis</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether to return barplot showing missingness.
The default is True.</p></li>
<li><p><strong>extra_vis</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return matrix plot showing missingness.
The default is False.</p></li>
<li><p><strong>save_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to folder where the results should be saved.
The default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If n_entries is incorrectly specified.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>miss_analysis gives a quick overview of the missingness of the provided
dataframe. You can provide the complete or prefiltered dataframe as input.
Providing n_entries allows you to specify how many of the entries of the dataframe
(sorted by missingness) are displayed (i.e. only display the n_entries columns with
most (or least) missing values) With the sort argument you can define
whether the dataframe is sorted by least to most missing values or vice versa
(using “descending” and “ascending”, respectively). The vis and extra_vis
arguments can be used to toggle the graphical output.
In case of large data (a lot of columns) those might be better turned off.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">miss_analysis</span><span class="p">(</span><span class="n">phos_expanded</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">twitchVsctrl</span><span class="o">+</span><span class="n">twitchVsmild</span><span class="o">+</span><span class="n">mildVsctrl</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">extra_vis</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-13.py">Source code</a>)</p>
<figure class="align-default" id="id7">
<img alt="_images/analysis-13_00.png" class="plot-directive" src="_images/analysis-13_00.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-13_00.png">png</a>, <a class="reference external" href=".//analysis-13_00.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-13_00.pdf">pdf</a>)</span><a class="headerlink" href="#id7" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="id8">
<img alt="_images/analysis-13_01.png" class="plot-directive" src="_images/analysis-13_01.png" />
<figcaption>
<p><span class="caption-text">(<a class="reference external" href=".//analysis-13_01.png">png</a>, <a class="reference external" href=".//analysis-13_01.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-13_01.pdf">pdf</a>)</span><a class="headerlink" href="#id8" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.rank_prod">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">rank_prod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#rank_prod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.rank_prod" title="Permalink to this definition">#</a></dt>
<dd><p>Perform RankProd test as in R RankProd package.</p>
<p>At the moment one sample test only.
Test for up and downregulated genes separatly therefore returns two p values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input dataframe.</p></li>
<li><p><strong>reps</strong> (<em>list of lists of str</em>) – Column names of the replicates.
Common replicates are grouped together in a single list.</p></li>
<li><p><strong>cond</strong> (<em>str</em><em>, </em><em>optional</em>) – Term to append to the newly generated colnames.
The default is “”.</p></li>
<li><p><strong>print_r</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to print the R output.
The default is False.</p></li>
<li><p><strong>correct_fc</strong> (<em>bool</em><em>, </em><em>optional</em>) – The rankProd package does not calculate fold-changes for rows with
missing values (p Values are calculated). If correct_fc is False
the original fold changes from rankProd are return, else fold
changes are calculated for all values after ignoring NaNs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – Input dataframe with additional columns from RankProd.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The adjusted p-values returned from the R backend are the percentage of
false postives calculated by the rankProd package. This is akin to corrected
p values, but care should be taken to name these values accordingly.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="autoprot.analysis.ttest">
<span class="sig-prename descclassname"><span class="pre">autoprot.analysis.</span></span><span class="sig-name descname"><span class="pre">ttest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust_p_vals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'two-sided'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/autoprot/analysis.html#ttest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#autoprot.analysis.ttest" title="Permalink to this definition">#</a></dt>
<dd><p>Perform one or two sample ttest.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input dataframe.</p></li>
<li><p><strong>reps</strong> (<em>list of str</em><em> or </em><em>list of lists of str</em>) – The replicates to be included in the statistical test.
Either a list of the replicates or a list containing two list with the
respective replicates.</p></li>
<li><p><strong>cond</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the condition.
This is used for naming the returned results.
The default is “”.</p></li>
<li><p><strong>return_fc</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate the fold-change of the provided data.
The processing of the fold-change can be controlled by the logged
switch.
The default is True.</p></li>
<li><p><strong>adjust_p_vals</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to adjust P-values. The default is True.</p></li>
<li><p><strong>alternative</strong> (<em>{'two-sided'</em><em>, </em><em>'less'</em><em>, </em><em>'greater'}</em><em>, </em><em>optional</em>) – <p>Defines the alternative hypothesis.
The following options are available (default is ‘two-sided’):</p>
<ul>
<li><p>’two-sided’: the mean of the underlying distribution of the sample
is different from the given population mean (<cite>popmean</cite>)</p></li>
<li><p>’less’: the mean of the underlying distribution of the sample is
less than the given population mean (<cite>popmean</cite>)</p></li>
<li><p>’greater’: the mean of the underlying distribution of the sample is
greater than the given population mean (<cite>popmean</cite>)</p></li>
</ul>
</p></li>
<li><p><strong>logged</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set to True if input values are log-transformed (or VSN normalised).
This returns the difference between values as log_fc, otherwise
values are log2 transformed to gain log_fc.
Default is true.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – Input dataframe with additional cols.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">twitchVsmild</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;log2_Ratio H/M normalized BC18_1&#39;</span><span class="p">,</span><span class="s1">&#39;log2_Ratio M/L normalized BC18_2&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio H/M normalized BC18_3&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio H/L normalized BC36_1&#39;</span><span class="p">,</span><span class="s1">&#39;log2_Ratio H/M normalized BC36_2&#39;</span><span class="p">,</span>
<span class="gp">... </span>                <span class="s1">&#39;log2_Ratio M/L normalized BC36_2&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protRatio</span> <span class="o">=</span> <span class="n">prot</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s2">&quot;Ratio .\/. normalized&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protLog</span> <span class="o">=</span> <span class="n">autoprot</span><span class="o">.</span><span class="n">preprocessing</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">prot</span><span class="p">,</span> <span class="n">protRatio</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_tt</span> <span class="o">=</span> <span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">ttest</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">protLog</span><span class="p">,</span> <span class="n">reps</span><span class="o">=</span><span class="n">twitchVsmild</span><span class="p">,</span> <span class="n">cond</span><span class="o">=</span><span class="s2">&quot;_TvM&quot;</span><span class="p">,</span> <span class="n">return_fc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">adjust_p_vals</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_tt</span><span class="p">[</span><span class="s2">&quot;pValue_TvM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-14.py">Source code</a>, <a class="reference external" href=".//analysis-14.png">png</a>, <a class="reference external" href=".//analysis-14.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-14.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-14.png" class="plot-directive" src="_images/analysis-14.png" />
</figure>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;a1&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                          <span class="s2">&quot;a2&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                          <span class="s2">&quot;a3&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                          <span class="s2">&quot;b1&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                          <span class="s2">&quot;b2&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),</span>
<span class="gp">... </span>                          <span class="s2">&quot;b3&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">4000</span><span class="p">),})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">autoprot</span><span class="o">.</span><span class="n">analysis</span><span class="o">.</span><span class="n">ttest</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">reps</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;a1&quot;</span><span class="p">,</span><span class="s2">&quot;a2&quot;</span><span class="p">,</span> <span class="s2">&quot;a3&quot;</span><span class="p">],[</span><span class="s2">&quot;b1&quot;</span><span class="p">,</span><span class="s2">&quot;b2&quot;</span><span class="p">,</span> <span class="s2">&quot;b3&quot;</span><span class="p">]])[</span><span class="s2">&quot;pValue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//analysis-15.py">Source code</a>, <a class="reference external" href=".//analysis-15.png">png</a>, <a class="reference external" href=".//analysis-15.hires.png">hires.png</a>, <a class="reference external" href=".//analysis-15.pdf">pdf</a>)</p>
<figure class="align-default">
<img alt="_images/analysis-15.png" class="plot-directive" src="_images/analysis-15.png" />
</figure>
</dd></dl>

</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="preprocessing.html" title="previous page">
      <i class="fas fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">Data Preprocessing</p>
      </div>
  </a>
  <a class='right-next' id="next-link" href="clustering.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Clustering</p>
  </div>
  <i class="fas fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
</div>

<div class="toc-item">
  
<div class="tocsection sourcelink">
    <a href="_sources/analysis.rst.txt">
        <i class="fas fa-file-alt"></i> Show Source
    </a>
</div>

</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=9b1a4fa89bdd0e95b23b"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2022, The autoprot contributors.<br>

</p>

  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.0.2.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>